<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Fast Unified Random Forests for Survival, Regression, and Classification (RF-SRC) — rfsrc • Random Forests for Survival, Regression, and Classification</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Fast Unified Random Forests for Survival, Regression, and Classification (RF-SRC) — rfsrc" />
<meta property="og:description" content="

Fast OpenMP parallel computing of random forests (Breiman 2001) for
  regression, classification, survival analysis (Ishwaran 2008),
  competing risks (Ishwaran 2012), multivariate (Segal and Xiao 2011),
  unsupervised (Mantero and Ishwaran 2020), quantile regression
  (Meinhausen 2006, Zhang et al. 2019), and class imbalanced
  q-classification (O'Brien and Ishwaran 2019).  Different splitting
  rules invoked under deterministic or random splitting (Geurts et
  al. 2006, Ishwaran 2015) are available for all families.  Variable
  importance (VIMP), and holdout VIMP, as well as confidence regions
  (Ishwaran and Lu 2019) can be calculated for single and grouped
  variables.  Fast interface for missing data imputation using a variety
  of different random forest methods (Tang and Ishwaran 2017).
Key items users should be aware of:
Visualize trees on your Safari or Google Chrome browser (works
    for all families).  See get.tree.
sidClustering (see sidClustering) for
    unsupervised data analysis.
Fast hold out VIMP (see holdout.vimp).
    Can be conservative but has good false discovery properties
    unlike Breiman-Cutler VIMP (see vimp).
VIMP confidence intervals using subsampling (see
    subsample).
q-classifier for class imbalanced data, including G-mean
    VIMP which is superior to Breiman-Cutler
    (see imbalanced).
Fast imputation using on the fly imputation,
    missForest and multivariate missForest (see
    impute).
Fast random forests using subsampling (see rfsrc.fast).

  
This is the main entry point to the randomForestSRC
  package.  For more information about this package and OpenMP parallel
  processing, use the command package?randomForestSRC." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=UA-000000-01"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-000000-01');
</script>


  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">Random Forests for Survival, Regression, and Classification</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Development version">2.11.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../articles/getstarted.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/getstarted.html">Getting Started</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/rfsrc-introduction.html">Introduction</a>
    </li>
    <li>
      <a href="../articles/installation.html">Installing randomForestSRC</a>
    </li>
    <li>
      <a href="../articles/parallel.html">Speedup Random Forest Analyses</a>
    </li>
    <li>
      <a href="../articles/regression.html">Regression Problem</a>
    </li>
    <li>
      <a href="../articles/classification.html">Classification Problem</a>
    </li>
    <li>
      <a href="../articles/survival.html">Survival Problem</a>
    </li>
    <li>
      <a href="../articles/imbalance.html">Classifiers for Class Imbalanced Data</a>
    </li>
    <li>
      <a href="../articles/rfsrc-subsample.html">Variable Importance with Subsampling Inferences</a>
    </li>
    <li>
      <a href="../articles/partial.html">Partial Plot</a>
    </li>
    <li>
      <a href="../articles/trees.html">Inferences from Trees</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Manual</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
<li>
  <a href="https://github.com/kogalur/randomForestSRC/issues">Bug Reporting</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/kogalur/randomForestSRC/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Fast Unified Random Forests for Survival, Regression, and Classification (RF-SRC)</h1>
    
    <div class="hidden name"><code>rfsrc.Rd</code></div>
    </div>

    <div class="ref-description">
    

<p>Fast OpenMP parallel computing of random forests (Breiman 2001) for
  regression, classification, survival analysis (Ishwaran 2008),
  competing risks (Ishwaran 2012), multivariate (Segal and Xiao 2011),
  unsupervised (Mantero and Ishwaran 2020), quantile regression
  (Meinhausen 2006, Zhang et al. 2019), and class imbalanced
  q-classification (O'Brien and Ishwaran 2019).  Different splitting
  rules invoked under deterministic or random splitting (Geurts et
  al. 2006, Ishwaran 2015) are available for all families.  Variable
  importance (VIMP), and holdout VIMP, as well as confidence regions
  (Ishwaran and Lu 2019) can be calculated for single and grouped
  variables.  Fast interface for missing data imputation using a variety
  of different random forest methods (Tang and Ishwaran 2017).</p>
<p>Key items users should be aware of:</p><ol>
<li><p>Visualize trees on your Safari or Google Chrome browser (works
    for all families).  See <code class='command'>get.tree</code>.</p></li>
<li><p>sidClustering (see <code class='command'><a href='sidClustering.rfsrc.html'>sidClustering</a></code>) for
    unsupervised data analysis.</p></li>
<li><p>Fast hold out VIMP (see <code class='command'><a href='holdout.vimp.rfsrc.html'>holdout.vimp</a></code>).
    Can be conservative but has good false discovery properties
    unlike Breiman-Cutler VIMP (see <code class='command'><a href='vimp.rfsrc.html'>vimp</a></code>).</p></li>
<li><p>VIMP confidence intervals using subsampling (see
    <code class='command'><a href='subsample.rfsrc.html'>subsample</a></code>).</p></li>
<li><p>q-classifier for class imbalanced data, including G-mean
    VIMP which is superior to Breiman-Cutler
    (see <code class='command'><a href='imbalanced.rfsrc.html'>imbalanced</a></code>).</p></li>
<li><p>Fast imputation using on the fly imputation,
    missForest and multivariate missForest (see
    <code class='command'><a href='impute.rfsrc.html'>impute</a></code>).</p></li>
<li><p>Fast random forests using subsampling (see <code class='command'><a href='rfsrc.fast.html'>rfsrc.fast</a></code>).</p></li>
</ol>
  
<p>This is the main entry point to the <span class="pkg">randomForestSRC</span>
  package.  For more information about this package and OpenMP parallel
  processing, use the command <code>package?randomForestSRC</code>.</p>
    </div>

    <pre class="usage"><span class='fu'>rfsrc</span><span class='op'>(</span><span class='va'>formula</span>, <span class='va'>data</span>, ntree <span class='op'>=</span> <span class='fl'>1000</span>,
  mtry <span class='op'>=</span> <span class='cn'>NULL</span>, ytry <span class='op'>=</span> <span class='cn'>NULL</span>,
  nodesize <span class='op'>=</span> <span class='cn'>NULL</span>, nodedepth <span class='op'>=</span> <span class='cn'>NULL</span>,
  splitrule <span class='op'>=</span> <span class='cn'>NULL</span>, nsplit <span class='op'>=</span> <span class='fl'>10</span>,
  importance <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='cn'>FALSE</span>, <span class='cn'>TRUE</span>, <span class='st'>"none"</span>, <span class='st'>"permute"</span>, <span class='st'>"random"</span>, <span class='st'>"anti"</span><span class='op'>)</span>,
  block.size <span class='op'>=</span> <span class='kw'>if</span> <span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/any.html'>any</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/sets.html'>is.element</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/character.html'>as.character</a></span><span class='op'>(</span><span class='va'>importance</span><span class='op'>)</span>,
                     <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"none"</span>, <span class='st'>"FALSE"</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span> <span class='cn'>NULL</span> <span class='kw'>else</span> <span class='fl'>10</span>,
  ensemble <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"all"</span>, <span class='st'>"oob"</span>, <span class='st'>"inbag"</span><span class='op'>)</span>,
  bootstrap <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"by.root"</span>, <span class='st'>"none"</span>, <span class='st'>"by.user"</span><span class='op'>)</span>,
  samptype <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"swor"</span>, <span class='st'>"swr"</span><span class='op'>)</span>, samp <span class='op'>=</span> <span class='cn'>NULL</span>, membership <span class='op'>=</span> <span class='cn'>FALSE</span>,
  sampsize <span class='op'>=</span> <span class='kw'>if</span> <span class='op'>(</span><span class='va'>samptype</span> <span class='op'>==</span> <span class='st'>"swor"</span><span class='op'>)</span> <span class='kw'>function</span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span><span class='op'>{</span><span class='va'>x</span> <span class='op'>*</span> <span class='fl'>.632</span><span class='op'>}</span> <span class='kw'>else</span> <span class='kw'>function</span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span><span class='op'>{</span><span class='va'>x</span><span class='op'>}</span>,
  na.action <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"na.omit"</span>, <span class='st'>"na.impute"</span><span class='op'>)</span>, nimpute <span class='op'>=</span> <span class='fl'>1</span>,
  <span class='va'>ntime</span>, <span class='va'>cause</span>,
  proximity <span class='op'>=</span> <span class='cn'>FALSE</span>, distance <span class='op'>=</span> <span class='cn'>FALSE</span>, forest.wt <span class='op'>=</span> <span class='cn'>FALSE</span>,
  xvar.wt <span class='op'>=</span> <span class='cn'>NULL</span>, yvar.wt <span class='op'>=</span> <span class='cn'>NULL</span>, split.wt <span class='op'>=</span> <span class='cn'>NULL</span>, case.wt  <span class='op'>=</span> <span class='cn'>NULL</span>,
  forest <span class='op'>=</span> <span class='cn'>TRUE</span>,
  var.used <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='cn'>FALSE</span>, <span class='st'>"all.trees"</span>, <span class='st'>"by.tree"</span><span class='op'>)</span>,
  split.depth <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='cn'>FALSE</span>, <span class='st'>"all.trees"</span>, <span class='st'>"by.tree"</span><span class='op'>)</span>,
  seed <span class='op'>=</span> <span class='cn'>NULL</span>,
  do.trace <span class='op'>=</span> <span class='cn'>FALSE</span>,
  statistics <span class='op'>=</span> <span class='cn'>FALSE</span>,
  <span class='va'>...</span><span class='op'>)</span>

<span class='co'>## convenient interface for growing a CART tree</span>
<span class='fu'>rfsrc.cart</span><span class='op'>(</span><span class='va'>formula</span>, <span class='va'>data</span>, ntree <span class='op'>=</span> <span class='fl'>1</span>, mtry <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/nrow.html'>ncol</a></span><span class='op'>(</span><span class='va'>data</span><span class='op'>)</span>, bootstrap <span class='op'>=</span> <span class='st'>"none"</span>, <span class='va'>...</span><span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>formula</th>
      <td><p>A symbolic description of the model to be fit.  If
    missing, unsupervised splitting is implemented.</p></td>
    </tr>
    <tr>
      <th>data</th>
      <td><p>Data frame containing the y-outcome and x-variables.</p></td>
    </tr>
    <tr>
      <th>ntree</th>
      <td><p>Number of trees.</p></td>
    </tr>
    <tr>
      <th>mtry</th>
      <td><p>Number of variables randomly selected as candidates for
    splitting a node.  The default is <code>p</code>/3 for regression,
    where <code>p</code> equals the number of variables.  For all
    other families (including unsupervised settings), the default is
    sqrt(<code>p</code>).  Values are always rounded up.</p></td>
    </tr>
    <tr>
      <th>ytry</th>
      <td><p>For unsupervised forests, sets the number of randomly
    selected pseudo-responses (see below for more details). 
    The default is <code>ytry</code>=1, which selects one pseudo-response.</p></td>
    </tr>
    <tr>
      <th>nodesize</th>
      <td><p>Minumum size of terminal node.
    The defaults are: survival (15), competing risk
    (15), regression (5), classification (1), mixed outcomes (3),
    unsupervised (3).  It is recommended to experiment with different
    <code>nodesize</code> values.</p></td>
    </tr>
    <tr>
      <th>nodedepth</th>
      <td><p>Maximum depth to which a tree should be grown.  The
    default behaviour is that this parameter is ignored.</p></td>
    </tr>
    <tr>
      <th>splitrule</th>
      <td><p>Splitting rule (see below).</p></td>
    </tr>
    <tr>
      <th>nsplit</th>
      <td><p>Non-negative integer value for number of random splits
   to consider for each candidate splitting variable.  This
   significantly increases speed.  When zero or NULL, uses much slower
   deterministic splitting where all possible splits considered.</p></td>
    </tr>
    <tr>
      <th>importance</th>
      <td><p>Method for computing variable importance (VIMP); see
    below.  Default action is <code>importance="none"</code> but VIMP can
    always be recovered later using <code class='command'>vimp</code> or
    <code class='command'>predict</code>.</p></td>
    </tr>
    <tr>
      <th>block.size</th>
      <td><p>Should the cumulative error rate be calculated on
    every tree?  When <code>NULL</code>, it will only be calculated on the
    last tree and the plot of the cumulative error rate will result in a
    flat line.  To view the cumulative error rate on every nth tree, set
    the value to an integer between <code>1</code> and <code>ntree</code>.  As an
    intended side effect, if importance is requested, VIMP is
    calculated in "blocks" of size equal to <code>block.size</code>, thus
    resulting in a useful compromise between ensemble and permutation
    VIMP. The default action is to use 10 trees.</p></td>
    </tr>
    <tr>
      <th>ensemble</th>
      <td><p>Specifies the type of ensemble.  By default both
    out-of-bag (OOB) and inbag ensembles are returned.  Always use OOB
    values for inference on the training data.</p></td>
    </tr>
    <tr>
      <th>bootstrap</th>
      <td><p>Bootstrap protocol.  The default is <code>by.root</code>
    which bootstraps the data by sampling with or without replacement
    (by default sampling is without replacement; see the option
    <code>samptype</code> below).  If <code>none</code> is chosen, the data is not
    bootstrapped at all.  If <code>by.user</code> is choosen, the bootstrap
    specified by <code>samp</code> is used.  It is not possible to return OOB
    ensembles or prediction error if <code>none</code> is in effect.</p></td>
    </tr>
    <tr>
      <th>samptype</th>
      <td><p>Type of bootstrap when <code>by.root</code> is in effect.
    Choices are <code>swor</code> (sampling without replacement) and
    <code>swr</code> (sampling with replacement).  Unlike Breiman's random
    forests, the default action here is sampling without replacement.
    Thus out-of-bag (OOB) technically means out-of-sample, but for
    legacy reasons we retain the term OOB.</p></td>
    </tr>
    <tr>
      <th>samp</th>
      <td><p>Bootstrap specification when <code>by.user</code> is in
    effect.  Array of dim <code>n x ntree</code> specifying how many
    times each record appears inbag in the bootstrap for each tree.</p></td>
    </tr>
    <tr>
      <th>membership</th>
      <td><p>Should terminal node membership and inbag
    information be returned?</p></td>
    </tr>
    <tr>
      <th>sampsize</th>
      <td><p>Function specifying size of bootstrap data when
    <code>by.root</code> is in effect.  For sampling without replacement, it is
    the requested size of the sample, which by default is .632 times the
    sample size.  For sampling with replacement, it is the sample size.
    Can also be specified by using a number.</p></td>
    </tr>
    <tr>
      <th>na.action</th>
      <td><p>Action taken if the data contains <code>NA</code>'s.
    Possible values are <code>na.omit</code> or <code>na.impute</code>.
    The default <code>na.omit</code> removes the entire record if
    even one of its entries is <code>NA</code> (for x-variables this applies only
    to those specifically listed in 'formula').  Selecting
    <code>na.impute</code> imputes the data.  Also see the function
    <code class='command'>impute</code> for fast direct imputation.</p></td>
    </tr>
    <tr>
      <th>nimpute</th>
      <td><p>Number of iterations of the missing data algorithm.
    Performance measures such as out-of-bag (OOB) error rates tend to
    become optimistic if <code>nimpute</code> is greater than 1.</p></td>
    </tr>
    <tr>
      <th>ntime</th>
      <td><p>Integer value used for survival to constrain ensemble
    calculations to a grid of <code>ntime</code> time points.  Alternatively
    if a vector of values of length greater than one is supplied, it is
    assumed these are the time points to be used to constrain the
    calculations (note that the constrained time points used will be the
    observed event times closest to the user supplied time points).  If
    no value is specified, the default action is to use all observed
    event times.</p></td>
    </tr>
    <tr>
      <th>cause</th>
      <td><p>Integer value between 1 and <code>J</code> indicating the event
    of interest for splitting a node for competing risks, where <code>J</code>
    is the number of event types. If not specified, the default is to
    use a composite splitting rule that averages over all event types.
    Can also be a vector of non-negative weights of length <code>J</code>
    specifying weights for each event (for example, passing a vector of
    ones reverts to the default composite split statistic).  Finally,
    regardless of how <code>cause</code> is specified, the returned forest
    object always provides estimates for all event types.</p></td>
    </tr>
    <tr>
      <th>proximity</th>
      <td><p>Proximity of cases as measured by the frequency of
    sharing the same terminal node.  This is an <code>n</code>x<code>n</code>
    matrix, which can be large.  Choices are <code>inbag</code>, <code>oob</code>,
    <code>all</code>, <code>TRUE</code>, or <code>FALSE</code>.  Setting <code>proximity =
      TRUE</code> is equivalent to <code>proximity = "inbag"</code>.</p></td>
    </tr>
    <tr>
      <th>distance</th>
      <td><p>Distance between cases as measured by the ratio of the
    sum of the count of edges from each case to their immediate common
    ancestor node to the sum of the count of edges from each case to the
    root node.  If the cases are co-terminal for a tree, this measure is
    zero and reduces to 1 - the proximity measure for these cases in a
    tree. This is an <code>n</code>x<code>n</code> matrix, which can be large.
    Choices are <code>inbag</code>, <code>oob</code>, <code>all</code>, <code>TRUE</code>,
    or <code>FALSE</code>.  Setting <code>distance = TRUE</code> is equivalent to
    <code>distance = "inbag"</code>.</p></td>
    </tr>
    <tr>
      <th>forest.wt</th>
      <td><p>Should the forest weight matrix be
    calculated?  Creates an <code>n</code>x<code>n</code> matrix which can be
    used for prediction and constructing customized estimators.  Choices
    are similar to proximity: <code>inbag</code>, <code>oob</code>, <code>all</code>,
    <code>TRUE</code>, or <code>FALSE</code>.  The default is <code>TRUE</code> which
    is equivalent to <code>inbag</code>.</p></td>
    </tr>
    <tr>
      <th>xvar.wt</th>
      <td><p>Vector of non-negative weights (does not have to sum
    to 1) representing the probability of selecting a variable for
    splitting.  Default is to use uniform weights.</p></td>
    </tr>
    <tr>
      <th>yvar.wt</th>
      <td><p>NOT YET IMPLEMENTED: Vector of non-negative weights
    (does not have to sum to 1) representing the probability of
    selecting a response as a candidate for the split statistic in
    unsupervised settings.  Default is to use uniform weights.</p></td>
    </tr>
    <tr>
      <th>split.wt</th>
      <td><p>Vector of non-negative weights used for multiplying
    the split statistic for a variable. A large value encourages the
    node to split on a specific variable. Default is to use uniform
    weights.</p></td>
    </tr>
    <tr>
      <th>case.wt</th>
      <td><p>Vector of non-negative weights (does not have to sum to
    1) for sampling cases.  Observations with larger weights will be
    selected with higher probability in the bootstrap (or subsampled)
    samples.  It is generally better to use real weights rather than
    integers. See the example below for the breast data set for an
    illustration of its use for class imbalanced data.</p></td>
    </tr>
    <tr>
      <th>forest</th>
      <td><p>Should the forest object be returned?  Used for
    prediction on new data and required by many of the package
    functions.</p></td>
    </tr>
    <tr>
      <th>var.used</th>
      <td><p>Return variables used for splitting?  Default is
    <code>FALSE</code>.  Possible values are <code>all.trees</code> which returns a
    vector of the number of times a split occurred on a variable, and
    <code>by.tree</code> which returns a matrix recording the number of times
    a split occurred on a variable in a specific tree.</p></td>
    </tr>
    <tr>
      <th>split.depth</th>
      <td><p>Records the minimal depth for each variable.
    Default is <code>FALSE</code>.  Possible values are <code>all.trees</code> which
    returns a matrix of the average minimal depth for a variable
    (columns) for a specific case (rows), and <code>by.tree</code> which
    returns a three-dimensional array recording minimal depth for a
    specific case (first dimension) for a variable (second dimension)
    for a specific tree (third dimension).</p></td>
    </tr>
    <tr>
      <th>seed</th>
      <td><p>Negative integer specifying seed for the random number
    generator.</p></td>
    </tr>
    <tr>
      <th>do.trace</th>
      <td><p>Number of seconds between updates to the user on
    approximate time to completion.</p></td>
    </tr>
    <tr>
      <th>statistics</th>
      <td><p>Should split statistics be returned?  Values can be
    parsed using <code class='command'>stat.split</code>.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Further arguments passed to or from other methods.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p></p><ol>
<li><p><em>Types of forests</em></p>
<p>There is no need to set the type of forest as the package
    automagically determines the underlying random forest requested from
    the type of response and the formula supplied.  There are several
    possible scenarios:</p>    
<p></p><ol>
<li><p>Regression forests for continuous responses.</p></li>
<li><p>Classification forests for factor responses.</p></li>
<li><p>Multivariate forests for continuous and/or factor responses
            and for mixed (both type) of responses.</p></li>
<li><p>Unsupervised forests when there is no response.</p></li>
<li><p>Survival forests for right-censored survival.</p></li>
<li><p>Competing risk survival forests for competing risk.</p></li>
</ol></li>
<li><p><em>Splitting</em></p>
<p></p><ol>
<li><p>Splitting rules are specified by the option <code>splitrule</code>.</p></li>
<li><p>For all families, pure random splitting can be invoked by setting
            <code>splitrule</code>="random".</p></li>
<li><p>For all families, computational speed can be increased using
            randomized splitting invoked by the option <code>nsplit</code>.
            See Improving Computational Speed.</p></li>
</ol></li>
<li><p><em>Available splitting rules</em></p>
<p></p><ul>
<li><p>Regression analysis:</p><ol>
<li><p><code>mse</code>: default split rule implements weighted
	mean-squared error splitting (Breiman et al. 1984, Chapter 8.4).</p></li>
<li><p><code>quantile.regr</code>: quantile regression splitting via
	the "check-loss" function.  Requires specifying the target
	quantiles.  See <code>quantreg.rfsrc</code> for further details.</p></li>
<li><p><code>la.quantile.regr</code>: local adaptive quantile
        regression splitting. See <code>quantreg.rfsrc</code>.</p></li>
</ol></li>
<li><p>Classification analysis:</p><ol>
<li><p><code>gini</code>: default splitrule implements Gini index
        splitting (Breiman et al. 1984, Chapter 4.3).</p></li>
<li><p><code>auc</code>: AUC (area under the ROC curve) splitting
	for both two-class and multiclass setttings.  AUC splitting is
	appropriate for imbalanced data.  See <code>imbalanced</code> for
	more information.</p></li>
<li><p><code>entropy</code>: entropy splitting (Breiman et
	al. 1984, Chapter 2.5, 4.3).</p></li>
</ol></li>
<li><p>Survival analysis:</p><ol>
<li><p><code>logrank</code>: default splitrule implements
	      log-rank splitting (Segal, 1988; Leblanc and Crowley, 1993).</p></li>
<li><p><code>bs.gradient</code>: gradient-based (global non-quantile)
	brier score splitting.  The time horizon used for the Brier
	score is set to the 90th percentile of the observed event times.
	This can be over-ridden by the option <code>prob</code>, which must be
	a value between 0 and 1 (set to .90 by default).</p></li>
<li><p><code>logrankscore</code>:  log-rank score splitting (Hothorn and
        Lausen, 2003).</p></li>
</ol></li>
<li><p>Competing risk analysis (for details see Ishwaran et al., 2014):</p><ol>
<li><p><code>logrankCR</code>: default splitrule implements a modified
	      weighted log-rank splitting rule modeled after Gray's test
	      (Gray, 1988).  Use this to find *all* variables 
	      that are informative and when the goal is long term
	      prediction.</p></li>
<li><p><code>logrank</code>: weighted log-rank splitting where each
	      event type is treated as the event of interest and all
	      other events are treated as censored.  The split rule is
	      the weighted value of each of log-rank statistics,
	      standardized by the variance.  Use this to find variables
	      that affect a *specific* cause of interest and when
	      the goal is a targeted analysis of a specific cause.
	      However in order for this to be effective, remember to set
	      the <code>cause</code> option to the targeted cause of interest.
	      See examples below.</p></li>
</ol></li>
<li><p>Multivariate analysis: When one or both regression and
	    classification responses are detected, a multivariate
	    normalized composite split rule of mean-squared error and Gini
	    index splitting is invoked.  See Tang and Ishwaran (2017)
	    for details on this splitting rule.  This splitting rule is
	    different than Segal and Xiao (2011) who use Mahalanobis
	    distance splitting.</p></li>
<li><p>Unsupervised analysis: In settings where there is no
            outcome, unsupervised splitting is invoked.  In this case,
            the mixed outcome splitting rule (above) is applied.  Also
            see <code class='command'>sidClustering</code> for a more sophisticated method for
            unsupervised analysis (Mantero and Ishwaran, 2020).</p></li>
<li><p>Custom splitting: All families except unsupervised are
            available for user defined custom splitting.  Some basic
            C-programming skills are required.  The harness for defining
            these rules is in <code>splitCustom.c</code>.  In this file we
            give examples of how to code rules for regression,
            classification, survival, and competing risk.  Each family
            can support up to sixteen custom split rules.  Specifying
            <code>splitrule="custom"</code> or <code>splitrule="custom1"</code> will
            trigger the first split rule for the family defined by the
            training data set. Multivariate families will need a custom
            split rule for both regression and classification.  In the
            examples, we demonstrate how the user is presented with the
            node specific membership.  The task is then to define a
            split statistic based on that membership.  Take note of the
            instructions in <code>splitCustom.c</code> on how to
            <em>register</em> the custom split rules.  It is suggested
            that the existing custom split rules be kept in place for
            reference and that the user proceed to develop
            <code>splitrule="custom2"</code> and so on. The package must be
            recompiled and installed for the custom split rules to
            become available.</p></li>
</ul></li>
<li><p><em>Improving computational speed</em></p>    
<p>See the function <code class='command'><a href='rfsrc.fast.html'>rfsrc.fast</a></code> for a fast
    implementation of <code>rfsrc</code>.  In general, the key methods for
    increasing speed are as follows:</p>    
<p></p><ul>
<li><p><em>Randomized splitting rules</em></p>
<p>Computational speed can be significantly improved with randomized
      splitting invoked by the option <code>nsplit</code>.  When <code>nsplit</code>
      is set to a non-zero positive integer, a maximum of <code>nsplit</code>
      split points are chosen randomly for each of the candidate
      splitting variables when splitting a tree node, thus significantly
      reducing the cost from having to consider all possible split-values.
      To revert to traditional deterministic splitting (all split
      values considered) use <code>nsplit</code>=0.</p>
<p>Randomized splitting for trees has a long history.  See for
      example, Loh and Shih (1997), Dietterich (2000), and Lin and Jeon
      (2006).  Geurts et al. (2006) recently introduced extremely
      randomized trees using what they call the extra-trees algorithm.
      We can see that this algorithm essentially corresponds to
      randomized splitting with <code>nsplit</code>=1.  In our experience
      however this is not always the optimal value for <code>nsplit</code> and
      may often be too low. See Ishwaran (2015).</p>
<p>Finally, for completely randomized (pure random) splitting use
      <code>splitrule</code>="random".  In pure splitting, nodes are split by
      randomly selecting a variable and randomly selecting its split
      point (Cutler and Zhao, 2001).</p></li>
<li><p><em>Subsampling</em></p>
<p>Subsampling can be used to reduce the size of the in-sample data
      used to grow a tree and therefore can greatly reduce computational
      load.  Subsampling is implemented using options <code>sampsize</code>
      and <code>samptype</code>.  See <code class='command'>rfsrc.fast</code> for a fast forest
      implementation using subsampling.</p></li>
<li><p><em>Unique time points</em></p>
<p>For large survival problems, users should consider setting
      <code>ntime</code> to a reasonably small value (such as 100 or 250).
      This constrains ensemble calculations such as survival functions
      to a restricted grid of time points.</p></li>
<li><p><em>Large number of variables</em></p>  
<p>The default setting <code>importance="none"</code> turns off variable
      importance (VIMP) calculations and considerably reduces
      computational times.  Variable importance can always
      be recovered later using functions <code class='command'>vimp</code> or
      <code class='command'>predict</code>.  Also consider using <code class='command'>max.subtree</code>
      which calculates minimal depth, a measure of the depth that a
      variable splits, and yields fast variable selection (Ishwaran,
      2010).</p></li>
</ul></li>
<li><p><em>Prediction Error</em></p>    
<p>Prediction error is calculated using OOB data.  Performance is
    measured in terms of mean-squared-error for regression, and
    misclassification error for classification.  A normalized Brier
    score (relative to a coin-toss) and the AUC (area under the ROC
    curve) is also provided upon printing a classification forest.
    Performance for Brier score can be specified using
    <code>perf.type="brier"</code>.  G-mean performance is also available, see
    the function <code class='command'>imbalanced</code> for more details.</p>    
<p>For survival, prediction error is measured by 1-C, where C is
    Harrell's (Harrell et al., 1982) concordance index.  Prediction error
    is between 0 and 1, and measures how well the predictor correctly
    ranks (classifies) two random individuals in terms of survival.  A
    value of 0.5 is no better than random guessing. A value of 0 is
    perfect.</p>    
<p>When bootstrapping is by <code>none</code>, a coherent OOB subset is not
    available to assess prediction error.  Thus, all outputs dependent
    on this are suppressed.  In such cases, prediction error is only
    available via classical cross-validation (the user will need to use
    the <code class='command'>predict.rfsrc</code> function).</p></li>
<li><p><em>Variable Importance (VIMP)</em></p>    
<p>To obtain VIMP use the option <code>importance</code>.  Setting this to
    "permute" or "TRUE" returns permutation VIMP from permuting OOB
    cases.  Choosing "random" replaces permutation with random
    assignment.  Thus when an OOB case is dropped down a tree, the case
    goes left or right randomly whenever a split is encountered
    for the target variable.  If "anti" is specified, the case is
    assigned to the opposite split.  By default <code>importance</code>="FALSE"
    and VIMP is not requested.</p>
<p>VIMP depends upon <code>block.size</code>, an integer value between 1 and
    <code>ntree</code>, specifying the number of trees in a block used to
    determine VIMP.  When <code>block.size</code>=1, VIMP is calculated by
    tree.  The difference between prediction error under the perturbed
    predictor and the original predictor is calculated for each tree and
    averaged over the forest.  This yields Breiman-Cutler VIMP (Breiman
    2001).</p>
<p>When <code>block.size</code>="ntree", VIMP is calculated across the forest
    by comparing the perturbed OOB forest ensemble (using all trees) to the
    unperturbed OOB forest ensemble (using all trees).  Unlike
    Breiman-Cutler VIMP, ensemble VIMP does not measure the tree average
    effect of <em>x</em>, but rather its overall forest effect.  This is
    called Ishwaran-Kogalur VIMP (Ishwaran et al. 2008).</p>
<p>A useful compromise between Breiman-Cutler (BC) and Ishwaran-Kogalur
    (IK) VIMP can be obtained by setting <code>block.size</code> to a value
    between 1 and <code>ntree</code>.  Smaller values are closer to BC and
    larger values closer to IK.  Smaller generally gives better
    accuracy, however computational times will be higher because VIMP is
    calculated over more blocks.  Also see <code class='command'>imbalanced</code> for
    imbalanced classification data where a larger <code>block.size</code>
    works better (O'Brien and Ishwaran, 2019).</p>
<p>See <code class='command'>vimp</code> for a user interface for extracting VIMP and
    <code class='command'>subsampling</code> for calculating confidence intervals for VIMP.</p>
<p>Also see <code class='command'>holdout.vimp</code> for holdout VIMP, which calculates
    importance by holding out variables.  This leads to a more
    conservative procedure but with good false discovery properties.</p>

<p>Note that for classification, VIMP is returned as a matrix with with
    J+1 columns where J is the number of classes.  The first column
    "all" is the unconditional VIMP, while the remaining columns are
    VIMP conditioned on cases corresponding to that class label.</p></li>
<li><p><em>Multivariate Forests</em></p>    
<p>Multivariate forests can be specified in two ways:</p>    
<p>rfsrc(Multivar(y1, y2, ..., yd) ~ . , my.data, ...)</p>    
<p>rfsrc(cbind(y1, y2, ..., yd) ~ . , my.data, ...)</p>
<p>A multivariate normalized composite splitting rule is used to split
    nodes.  The nature of the outcomes will inform the code as to the
    type of multivariate forest to be grown; i.e. whether it is real-valued,
    categorical, or a combination of both (mixed). Note that performance
    measures (when requested) are returned for all outcomes.</p></li>
<li><p><em>Unsupervised Forests and sidClustering</em></p>
<p>See sidClustering <code class='command'>sidClustering</code> for a more sophisticated
    method for unsupervised analysis.</p>
<p>Otherwise a more direct (but naive) way to proceed is to use the
    unsupervised splitting rule.  The following are equivalent ways to
    grow an unsupervised forest via unsupervised splitting:</p>    
<p>rfsrc(data = my.data)</p>    
<p>rfsrc(Unsupervised() ~ ., data = my.data)</p>    
<p>In unsupervised mode, features take turns acting as target
    y-outcomes and x-variables for splitting.  Specifically, <code>mtry</code>
    x-variables are randomly selected for splitting the node.  Then for
    each <code>mtry</code> feature, <code>ytry</code> variables are selected
    from the remaining features to act as the target pseduo-responses.
    Splitting uses the multivariate normalized composite splitting rule.</p>
<p>The default value of <code>ytry</code> is 1 but can be increased.  As
    illustration, the following equivalent unsupervised calls set
    <code>mtry=10</code> and <code>ytry</code>=5:</p>
<p>rfsrc(data = my.data, ytry = 5, mtry = 10)</p>    
<p>rfsrc(Unsupervised(5) ~ ., my.data, mtry = 10)</p>    
<p>Note that all performance values (error rates, VIMP, prediction) are
    turned off in unsupervised mode.</p></li>
<li><p><em>Survival, Competing Risks</em></p>    
<p></p><ol>
<li><p>Survival settings require a time and censoring variable which
      should be identifed in the formula as the response using the standard
      <code>Surv</code> formula specification.  A typical formula call looks like:</p>      
<p>Surv(my.time, my.status) ~ .</p>      
<p>where <code>my.time</code> and <code>my.status</code> are the variables names for
      the event time and status variable in the users data set.</p></li>
<li><p>For survival forests (Ishwaran et al. 2008), the censoring
      variable must be coded as a non-negative integer with 0 reserved for
      censoring and (usually) 1=death (event).</p></li>
<li><p>For competing risk forests (Ishwaran et al., 2013), the
      implementation is similar to survival, but with the following
      caveats:</p>      
<p></p><ul>
<li><p>Censoring must be coded as a non-negative integer, where 0
        indicates right-censoring, and non-zero values indicate different
        event types.  While 0,1,2,..,J is standard, and recommended,
        events can be coded non-sequentially, although 0 must always be used
        for censoring.</p></li>
<li><p>Setting the splitting rule to <code>logrankscore</code> will result
        in a survival analysis in which all events are treated as if they
        are the same type (indeed, they will coerced as such).</p></li>
<li><p>Generally, competing risks requires a larger <code>nodesize</code> than
        survival settings.</p></li>
</ul></li>
</ol></li>
<li><p><em>Missing data imputation</em></p>    
<p>Set <code>na.action="na.impute"</code> to impute missing data (both x and
    y-variables) using the missing data algorithm of Ishwaran et
    al. (2008).  However, see the function <code class='command'>impute</code> for an
    alternate way to do fast and accurate imputation.</p>    
<p>The missing data algorithm can be iterated by setting <code>nimpute</code>
    to a positive integer greater than 1.  When iterated, at the
    completion of each iteration, missing data is imputed using OOB
    non-missing terminal node data which is then used as input to grow a
    new forest.  A side effect of iteration is that missing values in
    the returned objects <code>xvar</code>, <code>yvar</code> are replaced by
    imputed values.  In other words the incoming data is overlaid with
    the missing data.  Also, performance measures such as error rates
    and VIMP become optimistically biased.</p>    
<p>Records in which all outcome and x-variable information are missing
    are removed from the forest analysis.  Variables having all missing
    values are also removed.</p></li>
<li><p><em>Allowable data types and factors</em></p>
<p>Data types must be real valued, integer, factor or logical --
    however all except factors are coerced and treated as if real
    valued.  For ordered x-variable factors, splits are similar to real
    valued variables.  For unordered factors, a split will move a subset
    of the levels in the parent node to the left daughter, and the
    complementary subset to the right daughter.  All possible
    complementary pairs are considered and apply to factors with an
    unlimited number of levels.  However, there is an optimization check
    to ensure number of splits attempted is not greater than number of
    cases in a node or the value of <code>nsplit</code>.</p>
<p>For coherence, an immutable map is applied to each factor that
    ensures factor levels in the training data are consistent with the
    factor levels in any subsequent test data.  This map is applied to
    each factor before and after the native C library is executed.
    Because of this, if all x-variables all factors, then computational
    time will be long in high dimensional problems.  Consider converting
    factors to real if this is the case.</p></li>
</ol>
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>An object of class <code>(rfsrc, grow)</code> with the following
  components:</p>  
<p></p>
<dt>call</dt><dd><p>The original call to <code class='command'>rfsrc</code>.</p></dd>
  <dt>family</dt><dd><p>The family used in the analysis.</p></dd>
  <dt>n</dt><dd><p>Sample size of the data (depends upon <code>NA</code>'s, see <code>na.action</code>).</p></dd>
  <dt>ntree</dt><dd><p>Number of trees grown.</p></dd>
  <dt>mtry</dt><dd><p>Number of variables randomly selected for splitting at each node.</p></dd>
  <dt>nodesize</dt><dd><p>Minimum size of terminal nodes.</p></dd>
  <dt>nodedepth</dt><dd><p>Maximum depth allowed for a tree.</p></dd>
  <dt>splitrule</dt><dd><p>Splitting rule used.</p></dd>
  <dt>nsplit</dt><dd><p>Number of randomly selected split points.</p></dd>
  <dt>yvar</dt><dd><p>y-outcome values.</p></dd>
  <dt>yvar.names</dt><dd><p>A character vector of the y-outcome names.</p></dd>
  <dt>xvar</dt><dd><p>Data frame of x-variables.</p></dd>
  <dt>xvar.names</dt><dd><p>A character vector of the x-variable names.</p></dd>
  <dt>xvar.wt</dt><dd><p>Vector of non-negative weights specifying the
    probability used to select a variable for splitting a node.</p></dd>
  <dt>split.wt</dt><dd><p>Vector of non-negative weights specifying
    multiplier by which the split statistic for a covariate is adjusted.</p></dd>
  <dt>cause.wt</dt><dd><p>Vector of weights used for the composite competing
    risk splitting rule.</p></dd>
  <dt>leaf.count</dt><dd><p>Number of terminal nodes for each tree in the
    forest. Vector of length <code>ntree</code>.  A value of zero indicates
    a rejected tree (can occur when imputing missing data).
    Values of one indicate tree stumps.</p></dd>
  <dt>proximity</dt><dd><p>Proximity matrix recording the frequency of pairs of data points
    occur within the same terminal node.</p></dd>
  <dt>forest</dt><dd><p>If <code>forest=TRUE</code>, the forest object is returned.
    This object is used for prediction with new test data
	  sets and is required for other R-wrappers.</p></dd>
  <dt>forest.wt</dt><dd><p>Forest weight matrix.</p></dd>
  <dt>membership</dt><dd><p>Matrix recording terminal node membership where
    each column records node mebership for a case for a tree (rows).</p></dd>
  <dt>splitrule</dt><dd><p>Splitting rule used.</p></dd>
  <dt>inbag</dt><dd><p>Matrix recording inbag membership where each column
    contains the number of times that a case appears in the bootstrap
    sample for a tree (rows).</p></dd>
  <dt>var.used</dt><dd><p>Count of the number of times a variable is used in
    growing the forest.</p></dd>
  <dt>imputed.indv</dt><dd><p>Vector of indices for cases with missing
    values.</p></dd>
  <dt>imputed.data</dt><dd><p>Data frame of the imputed data. The first
	  column(s) are reserved for the y-responses, after which the
	  x-variables are listed.</p></dd>
  <dt>split.depth</dt><dd><p>Matrix [i][j] or array [i][j][k] recording the
    minimal depth for variable [j] for case [i], either averaged over
    the forest, or by tree [k].</p></dd>
  <dt>node.stats</dt><dd><p>Split statistics returned when
    <code>statistics=TRUE</code> which can be parsed using <code class='command'>stat.split</code>.</p></dd>

  <dt>err.rate</dt><dd><p>Tree cumulative OOB error rate.</p></dd>
  <dt>err.block.rate</dt><dd><p>When <code>importance=TRUE</code>, vector of the
  cumulative error rate for each ensemble block comprised of
  <code>block.size</code> trees.  So with <code>block.size = 10</code>, entries are
  the cumulative error rate for the first 10 trees, the first 20 trees,
  30 trees, and so on.  As another exmple, if <code>block.size = 1</code>,
  entries are the error rate for each tree.</p></dd>

  
    
  <dt>importance</dt><dd><p>Variable importance (VIMP) for each x-variable.</p></dd>
  <dt>predicted</dt><dd><p>In-bag predicted value.</p></dd>
  <dt>predicted.oob</dt><dd><p>OOB predicted value.</p></dd>
  
  <dt>++++++++</dt><dd><p>for classification settings, additionally ++++++++</p></dd> 
  <dt>class</dt><dd><p>In-bag predicted class labels.</p></dd>
  <dt>class.oob</dt><dd><p>OOB predicted class labels.</p></dd>
  
  
  <dt>++++++++</dt><dd><p>for multivariate settings, additionally ++++++++</p></dd> 
  
  <dt>regrOutput</dt><dd><p>List containing performance values for
	  multivariate regression responses (applies only in
	  multivariate settings).</p></dd>
	<dt>clasOutput</dt><dd><p>List containing performance values for
	  multivariate categorical (factor) responses (applies only in
	  multivariate settings).</p></dd>
  
  <dt>++++++++</dt><dd><p>for survival settings, additionally ++++++++</p></dd> 
  
  <dt>survival</dt><dd><p>In-bag survival function.</p></dd>
  <dt>survival.oob</dt><dd><p>OOB survival function.</p></dd>
  <dt>chf</dt><dd><p>In-bag cumulative hazard function (CHF).</p></dd>
  <dt>chf.oob</dt><dd><p>OOB CHF.</p></dd>
  <dt>time.interest</dt><dd><p>Ordered unique death times.</p></dd>
  <dt>ndead</dt><dd><p>Number of deaths.</p></dd>
  
  <dt>++++++++</dt><dd><p>for competing risks, additionally ++++++++</p></dd> 
  
  <dt>chf</dt><dd><p>In-bag cause-specific cumulative hazard function (CSCHF)
    for each event.</p></dd>
  <dt>chf.oob</dt><dd><p>OOB CSCHF.</p></dd>
  <dt>cif</dt><dd><p>In-bag cumulative incidence function (CIF) for each event.</p></dd>
  <dt>cif.oob</dt><dd><p>OOB CIF.</p></dd>
  <dt>time.interest</dt><dd><p>Ordered unique event times.</p></dd>
  <dt>ndead</dt><dd><p>Number of events.</p></dd>

    <h2 class="hasAnchor" id="note"><a class="anchor" href="#note"></a>Note</h2>

    

<p>Values returned depend heavily on the family.  In particular,
  predicted values from the forest (<code>predicted</code> and
  <code>predicted.oob</code>) are as follows:</p>    
<p></p><ol>
<li><p>For regression, a vector of predicted y-responses.</p></li>
<li><p>For classification, a matrix with columns containing the
      estimated class probability for each class.  Performance values
      and VIMP for classification are reported as a matrix with J+1
      columns where J is the number of classes.  The first column "all"
      is the unconditional value for performance (VIMP), while the
      remaining columns are performance (VIMP) conditioned on cases
      corresponding to that class label.</p></li>
<li><p>For survival, a vector of mortality values (Ishwaran et al.,
      2008) representing estimated risk for each individual calibrated
      to the scale of the number of events (as a specific example, if
      <em>i</em> has a mortality value of 100, then if all individuals had
      the same x-values as <em>i</em>, we would expect an average of 100
      events).  Also returned are matrices containing
      the CHF and survival function.  Each row corresponds to an
      individual's ensemble CHF or survival function evaluated at each
      time point in <code>time.interest</code>.</p></li>
<li><p>For competing risks, a matrix with one column for each event
      recording the expected number of life years lost due to the event
      specific cause up to the maximum follow up (Ishwaran et al.,
      2013).  Also returned are the cause-specific cumulative hazard
      function (CSCHF) and the cumulative incidence function (CIF) for
      each event type.  These are encoded as a three-dimensional array,
      with the third dimension used for the event type, each time point
      in <code>time.interest</code> making up the second dimension (columns),
      and the case (individual) being the first dimension (rows).</p></li>
<li><p>For multivariate families, predicted values (and other
      performance values such as VIMP and error rates) are stored in the
      lists <code>regrOutput</code> and <code>clasOutput</code> which can be
      extracted using functions <code>get.mv.error</code>,
      <code>get.mv.predicted</code> and <code>get.mv.vimp</code>.</p></li>
</ol><p></p>
    <h2 class="hasAnchor" id="author"><a class="anchor" href="#author"></a>Author</h2>

    <p>Hemant Ishwaran and Udaya B. Kogalur</p>
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>Breiman L., Friedman J.H., Olshen R.A. and Stone C.J. (1984).
  <em>Classification and Regression Trees</em>, Belmont, California.</p>  
<p>Breiman L. (2001). Random forests, <em>Machine Learning</em>, 45:5-32.</p>  
<p>Cutler A. and Zhao G. (2001). PERT-Perfect random tree ensembles.
  <em>Comp. Sci. Statist.</em>, 33: 490-497.</p>
<p>Dietterich, T. G. (2000). An experimental comparison of three methods for
  constructing ensembles of decision trees: bagging, boosting, and randomization.
  <em>Machine Learning</em>, 40, 139-157.</p>  
<p>Gray R.J. (1988).  A class of k-sample tests for comparing the
  cumulative incidence of a competing risk, <em>Ann. Statist.</em>,
  16: 1141-1154.</p>  
<p>Geurts, P., Ernst, D. and Wehenkel, L., (2006). Extremely randomized
  trees. <em>Machine learning</em>, 63(1):3-42.</p>
<p>Harrell et al. F.E. (1982).  Evaluating the yield of medical tests,
  <em>J. Amer. Med. Assoc.</em>, 247:2543-2546.</p>  
<p>Hothorn T. and Lausen B. (2003). On the exact distribution of maximally selected
  rank statistics, <em>Comp. Statist. Data Anal.</em>, 43:121-137.</p>  
<p>Ishwaran H. (2007).  Variable importance in binary regression
  trees and forests,  <em>Electronic J. Statist.</em>, 1:519-537.</p>  
<p>Ishwaran H. and Kogalur U.B. (2007).  Random survival forests for R,
  <em>Rnews</em>, 7(2):25-31.</p>  
<p>Ishwaran H., Kogalur U.B., Blackstone E.H. and Lauer M.S.
  (2008).  Random survival forests, <em>Ann. App.
    Statist.</em>, 2:841-860.</p>  
<p>Ishwaran H., Kogalur U.B., Gorodeski E.Z, Minn A.J. and
  Lauer M.S. (2010).  High-dimensional variable selection for survival
  data.  <em>J. Amer. Statist. Assoc.</em>, 105:205-217.</p>  
<p>Ishwaran H., Kogalur U.B., Chen X. and Minn A.J. (2011). Random survival
  forests for high-dimensional data. <em>Stat. Anal. Data Mining</em>, 4:115-132</p>  
<p>Ishwaran H., Gerds T.A., Kogalur U.B., Moore R.D., Gange S.J. and Lau
  B.M. (2014). Random survival forests for competing risks.
  <em>Biostatistics</em>, 15(4):757-773.</p>  
<p>Ishwaran H. and Malley J.D. (2014). Synthetic learning
  machines. <em>BioData Mining</em>, 7:28.</p>
<p>Ishwaran H. (2015).  The effect of splitting on random forests.
  <em>Machine Learning</em>, 99:75-118.</p>
<p>Lin, Y. and Jeon, Y. (2006). Random forests and adaptive nearest
  neighbors. <em>J. Amer. Statist. Assoc.</em>, 101(474), 578-590.</p>  
<p>Lu M., Sadiq S., Feaster D.J. and Ishwaran H. (2018). Estimating
  individual treatment effect in observational data using random forest
  methods. <em>J. Comp. Graph. Statist</em>, 27(1), 209-219</p>
<p>Ishwaran H. and Lu M.  (2019).  Standard errors and confidence
  intervals for variable importance in random forest regression,
  classification, and survival. <em>Statistics in Medicine</em>, 38,
  558-582.</p>  
<p>LeBlanc M. and Crowley J. (1993).  Survival trees by goodness of split,
  <em>J. Amer. Statist. Assoc.</em>, 88:457-467.</p>  
<p>Loh W.-Y and Shih Y.-S (1997).  Split selection methods for
  classification trees, <em>Statist. Sinica</em>, 7:815-840.</p>
<p>Mantero A. and Ishwaran H. (2020).  Unsupervised random forests.
  To appear in <em>Statistical Analysis and Data Mining</em>.</p>
<p>Meinshausen N. (2006) Quantile regression forests, <em>Journal of
  Machine Learning Research</em>, 7:983-999.</p>
<p>Mogensen, U.B, Ishwaran H. and Gerds T.A. (2012). Evaluating random
  forests for survival analysis using prediction error curves,
  <em>J.  Statist. Software</em>, 50(11): 1-23.</p>  
<p>O'Brien R. and Ishwaran H. (2019).  A random forests quantile
  classifier for class imbalanced data. <em>Pattern Recognition</em>,
  90, 232-249</p>
<p>Segal M.R. (1988).  Regression trees for censored data,
  <em>Biometrics</em>,  44:35-47.</p>
<p>Segal M.R. and Xiao Y. Multivariate random
  forests. (2011). <em>Wiley Interdisciplinary Reviews: Data Mining
  and Knowledge Discovery</em>. 1(1):80-87.</p>  
<p>Tang F. and Ishwaran H. (2017).  Random forest missing data
  algorithms.  <em>Statistical Analysis and Data Mining</em>, 10:363-377.</p>
<p>Zhang H., Zimmerman J., Nettleton D. and Nordman D.J. (2019).  Random
  forest prediction intervals. <em>The American Statistician</em>. 4:1-5.</p>  


    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code class='command'><a href='find.interaction.rfsrc.html'>find.interaction.rfsrc</a></code>,</p>
<p><code class='command'>get.tree.rfsrc</code>,</p>
<p><code class='command'><a href='holdout.vimp.rfsrc.html'>holdout.vimp.rfsrc</a></code>,</p>
<p><code class='command'><a href='imbalanced.rfsrc.html'>imbalanced.rfsrc</a></code>,
  <code class='command'><a href='impute.rfsrc.html'>impute.rfsrc</a></code>,</p>
<p><code class='command'>max.subtree.rfsrc</code>,</p>
<p><code class='command'><a href='partial.rfsrc.html'>partial.rfsrc</a></code>,
  <code class='command'><a href='plot.competing.risk.rfsrc.html'>plot.competing.risk.rfsrc</a></code>,
  <code class='command'><a href='plot.rfsrc.html'>plot.rfsrc</a></code>,
  <code class='command'><a href='plot.survival.rfsrc.html'>plot.survival.rfsrc</a></code>,
  <code class='command'><a href='plot.variable.rfsrc.html'>plot.variable.rfsrc</a></code>,
  <code class='command'><a href='predict.rfsrc.html'>predict.rfsrc</a></code>,
  <code class='command'><a href='print.rfsrc.html'>print.rfsrc</a></code>,</p>
<p><code class='command'><a href='quantreg.rfsrc.html'>quantreg.rfsrc</a></code>,</p>
<p><code class='command'><a href='rfsrc.fast.html'>rfsrc.fast</a></code>,</p>  
<p><code class='command'><a href='sidClustering.rfsrc.html'>sidClustering.rfsrc</a></code>,</p>
<p><code class='command'><a href='stat.split.rfsrc.html'>stat.split.rfsrc</a></code>,
  <code class='command'><a href='subsample.rfsrc.html'>subsample.rfsrc</a></code>,
  <code class='command'><a href='synthetic.html'>synthetic.rfsrc</a></code>,</p>  
<p><code class='command'><a href='tune.rfsrc.html'>tune.rfsrc</a></code>,</p>
<p><code class='command'><a href='var.select.rfsrc.html'>var.select.rfsrc</a></code>,
  <code class='command'><a href='vimp.rfsrc.html'>vimp.rfsrc</a></code></p></div>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><span class='co'>##------------------------------------------------------------</span>
<span class='co'>## survival analysis</span>
<span class='co'>##------------------------------------------------------------</span>

<span class='co'>## veteran data</span>
<span class='co'>## randomized trial of two treatment regimens for lung cancer</span>
<span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span><span class='op'>(</span><span class='va'>veteran</span>, package <span class='op'>=</span> <span class='st'>"randomForestSRC"</span><span class='op'>)</span>
<span class='va'>v.obj</span> <span class='op'>&lt;-</span> <span class='fu'>rfsrc</span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/pkg/survival/man/Surv.html'>Surv</a></span><span class='op'>(</span><span class='va'>time</span>, <span class='va'>status</span><span class='op'>)</span> <span class='op'>~</span> <span class='va'>.</span>, data <span class='op'>=</span> <span class='va'>veteran</span>, 
                   ntree <span class='op'>=</span> <span class='fl'>100</span>, block.size <span class='op'>=</span> <span class='fl'>1</span><span class='op'>)</span>

<span class='co'>## print tree number 3</span>
<span class='fu'><a href='https://rdrr.io/r/graphics/plot.default.html'>plot</a></span><span class='op'>(</span><span class='fu'>get.tree</span><span class='op'>(</span><span class='va'>v.obj</span>, <span class='fl'>3</span><span class='op'>)</span><span class='op'>)</span>

<span class='co'>## print the grow object and plot various useful details</span>
<span class='fu'><a href='https://rdrr.io/r/base/print.html'>print</a></span><span class='op'>(</span><span class='va'>v.obj</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/graphics/plot.default.html'>plot</a></span><span class='op'>(</span><span class='va'>v.obj</span><span class='op'>)</span>

<span class='co'>## plot survival curves for first 10 individuals -- direct way</span>
<span class='fu'><a href='https://rdrr.io/r/graphics/matplot.html'>matplot</a></span><span class='op'>(</span><span class='va'>v.obj</span><span class='op'>$</span><span class='va'>time.interest</span>, <span class='fl'>100</span> <span class='op'>*</span> <span class='fu'><a href='https://rdrr.io/r/base/t.html'>t</a></span><span class='op'>(</span><span class='va'>v.obj</span><span class='op'>$</span><span class='va'>survival.oob</span><span class='op'>[</span><span class='fl'>1</span><span class='op'>:</span><span class='fl'>10</span>, <span class='op'>]</span><span class='op'>)</span>,
    xlab <span class='op'>=</span> <span class='st'>"Time"</span>, ylab <span class='op'>=</span> <span class='st'>"Survival"</span>, type <span class='op'>=</span> <span class='st'>"l"</span>, lty <span class='op'>=</span> <span class='fl'>1</span><span class='op'>)</span>

<span class='co'>## plot survival curves for first 10 individuals</span>
<span class='co'>## using function "plot.survival" </span>
<span class='fu'><a href='plot.survival.rfsrc.html'>plot.survival</a></span><span class='op'>(</span><span class='va'>v.obj</span>, subset <span class='op'>=</span> <span class='fl'>1</span><span class='op'>:</span><span class='fl'>10</span><span class='op'>)</span>

<span class='co'># \donttest{</span>
<span class='co'>## fast nodesize optimization for veteran data</span>
<span class='co'>## optimal nodesize in survival is larger than other families</span>
<span class='co'>## see the function "tune" for more examples</span>
<span class='fu'><a href='tune.rfsrc.html'>tune.nodesize</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/pkg/survival/man/Surv.html'>Surv</a></span><span class='op'>(</span><span class='va'>time</span>,<span class='va'>status</span><span class='op'>)</span> <span class='op'>~</span> <span class='va'>.</span>, <span class='va'>veteran</span><span class='op'>)</span>


<span class='co'>## Primary biliary cirrhosis (PBC) of the liver</span>
<span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span><span class='op'>(</span><span class='va'>pbc</span>, package <span class='op'>=</span> <span class='st'>"randomForestSRC"</span><span class='op'>)</span>
<span class='va'>pbc.obj</span> <span class='op'>&lt;-</span> <span class='fu'>rfsrc</span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/pkg/survival/man/Surv.html'>Surv</a></span><span class='op'>(</span><span class='va'>days</span>, <span class='va'>status</span><span class='op'>)</span> <span class='op'>~</span> <span class='va'>.</span>, <span class='va'>pbc</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/print.html'>print</a></span><span class='op'>(</span><span class='va'>pbc.obj</span><span class='op'>)</span>

<span class='co'>##------------------------------------------------------------</span>
<span class='co'>## trees can be plotted for any family </span>
<span class='co'>## see get.tree for details and more examples</span>
<span class='co'>##------------------------------------------------------------</span>

<span class='co'>## survival where factors have many levels</span>
<span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span><span class='op'>(</span><span class='va'>veteran</span>, package <span class='op'>=</span> <span class='st'>"randomForestSRC"</span><span class='op'>)</span>
<span class='va'>vd</span> <span class='op'>&lt;-</span> <span class='va'>veteran</span>
<span class='va'>vd</span><span class='op'>$</span><span class='va'>celltype</span><span class='op'>=</span><span class='fu'><a href='https://rdrr.io/r/base/factor.html'>factor</a></span><span class='op'>(</span><span class='va'>vd</span><span class='op'>$</span><span class='va'>celltype</span><span class='op'>)</span>
<span class='va'>vd</span><span class='op'>$</span><span class='va'>diagtime</span><span class='op'>=</span><span class='fu'><a href='https://rdrr.io/r/base/factor.html'>factor</a></span><span class='op'>(</span><span class='va'>vd</span><span class='op'>$</span><span class='va'>diagtime</span><span class='op'>)</span>
<span class='va'>vd.obj</span> <span class='op'>&lt;-</span> <span class='fu'>rfsrc</span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/pkg/survival/man/Surv.html'>Surv</a></span><span class='op'>(</span><span class='va'>time</span>,<span class='va'>status</span><span class='op'>)</span><span class='op'>~</span><span class='va'>.</span>, <span class='va'>vd</span>, ntree <span class='op'>=</span> <span class='fl'>100</span>, nodesize <span class='op'>=</span> <span class='fl'>5</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/graphics/plot.default.html'>plot</a></span><span class='op'>(</span><span class='fu'>get.tree</span><span class='op'>(</span><span class='va'>vd.obj</span>, <span class='fl'>3</span><span class='op'>)</span><span class='op'>)</span>

<span class='co'>## classification</span>
<span class='va'>iris.obj</span> <span class='op'>&lt;-</span> <span class='fu'>rfsrc</span><span class='op'>(</span><span class='va'>Species</span> <span class='op'>~</span><span class='va'>.</span>, data <span class='op'>=</span> <span class='va'>iris</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/graphics/plot.default.html'>plot</a></span><span class='op'>(</span><span class='fu'>get.tree</span><span class='op'>(</span><span class='va'>iris.obj</span>, <span class='fl'>25</span>, class.type <span class='op'>=</span> <span class='st'>"bayes"</span><span class='op'>)</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/graphics/plot.default.html'>plot</a></span><span class='op'>(</span><span class='fu'>get.tree</span><span class='op'>(</span><span class='va'>iris.obj</span>, <span class='fl'>25</span>, target <span class='op'>=</span> <span class='st'>"setosa"</span><span class='op'>)</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/graphics/plot.default.html'>plot</a></span><span class='op'>(</span><span class='fu'>get.tree</span><span class='op'>(</span><span class='va'>iris.obj</span>, <span class='fl'>25</span>, target <span class='op'>=</span> <span class='st'>"versicolor"</span><span class='op'>)</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/graphics/plot.default.html'>plot</a></span><span class='op'>(</span><span class='fu'>get.tree</span><span class='op'>(</span><span class='va'>iris.obj</span>, <span class='fl'>25</span>, target <span class='op'>=</span> <span class='st'>"virginica"</span><span class='op'>)</span><span class='op'>)</span>

<span class='co'>## ------------------------------------------------------------</span>
<span class='co'>## simple example of VIMP using iris classification</span>
<span class='co'>## ------------------------------------------------------------</span>

<span class='co'>## directly from grow call </span>
<span class='fu'><a href='https://rdrr.io/r/base/print.html'>print</a></span><span class='op'>(</span><span class='fu'>rfsrc</span><span class='op'>(</span><span class='va'>Species</span><span class='op'>~</span><span class='va'>.</span>,<span class='va'>iris</span>,importance<span class='op'>=</span><span class='cn'>TRUE</span><span class='op'>)</span><span class='op'>$</span><span class='va'>importance</span><span class='op'>)</span>

<span class='co'>## note by default VIMP (and performance) uses misclassification error,</span>
<span class='co'>## but brier prediction error can be specified</span>
<span class='fu'><a href='https://rdrr.io/r/base/print.html'>print</a></span><span class='op'>(</span><span class='fu'>rfsrc</span><span class='op'>(</span><span class='va'>Species</span><span class='op'>~</span><span class='va'>.</span>,<span class='va'>iris</span>,importance<span class='op'>=</span><span class='cn'>TRUE</span>,perf.type<span class='op'>=</span><span class='st'>"brier"</span><span class='op'>)</span><span class='op'>$</span><span class='va'>importance</span><span class='op'>)</span>

<span class='co'>## using vimp function (see vimp help file for details)</span>
<span class='va'>iris.obj</span> <span class='op'>&lt;-</span> <span class='fu'>rfsrc</span><span class='op'>(</span><span class='va'>Species</span> <span class='op'>~</span><span class='va'>.</span>, data <span class='op'>=</span> <span class='va'>iris</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/print.html'>print</a></span><span class='op'>(</span><span class='fu'><a href='vimp.rfsrc.html'>vimp</a></span><span class='op'>(</span><span class='va'>iris.obj</span><span class='op'>)</span><span class='op'>$</span><span class='va'>importance</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/print.html'>print</a></span><span class='op'>(</span><span class='fu'><a href='vimp.rfsrc.html'>vimp</a></span><span class='op'>(</span><span class='va'>iris.obj</span>,perf.type<span class='op'>=</span><span class='st'>"brier"</span><span class='op'>)</span><span class='op'>$</span><span class='va'>importance</span><span class='op'>)</span>

<span class='co'>## hold out vimp (see holdout.vimp help file for details)</span>
<span class='fu'><a href='https://rdrr.io/r/base/print.html'>print</a></span><span class='op'>(</span><span class='fu'><a href='holdout.vimp.rfsrc.html'>holdout.vimp</a></span><span class='op'>(</span><span class='va'>Species</span><span class='op'>~</span><span class='va'>.</span>,<span class='va'>iris</span><span class='op'>)</span><span class='op'>$</span><span class='va'>importance</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/print.html'>print</a></span><span class='op'>(</span><span class='fu'><a href='holdout.vimp.rfsrc.html'>holdout.vimp</a></span><span class='op'>(</span><span class='va'>Species</span><span class='op'>~</span><span class='va'>.</span>,<span class='va'>iris</span>,perf.type<span class='op'>=</span><span class='st'>"brier"</span><span class='op'>)</span><span class='op'>$</span><span class='va'>importance</span><span class='op'>)</span>

<span class='co'>## ------------------------------------------------------------</span>
<span class='co'>## confidence interval for vimp using subsampling</span>
<span class='co'>## compare with holdout vimp</span>
<span class='co'>## ------------------------------------------------------------</span>

<span class='co'>## new York air quality measurements</span>
<span class='va'>o</span> <span class='op'>&lt;-</span> <span class='fu'>rfsrc</span><span class='op'>(</span><span class='va'>Ozone</span> <span class='op'>~</span> <span class='va'>.</span>, data <span class='op'>=</span> <span class='va'>airquality</span><span class='op'>)</span>
<span class='va'>so</span> <span class='op'>&lt;-</span> <span class='fu'><a href='subsample.rfsrc.html'>subsample</a></span><span class='op'>(</span><span class='va'>o</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/graphics/plot.default.html'>plot</a></span><span class='op'>(</span><span class='va'>so</span><span class='op'>)</span>

<span class='co'>## compare with holdout vimp</span>
<span class='fu'><a href='https://rdrr.io/r/base/print.html'>print</a></span><span class='op'>(</span><span class='fu'><a href='holdout.vimp.rfsrc.html'>holdout.vimp</a></span><span class='op'>(</span><span class='va'>Ozone</span> <span class='op'>~</span> <span class='va'>.</span>, data <span class='op'>=</span> <span class='va'>airquality</span><span class='op'>)</span><span class='op'>$</span><span class='va'>importance</span><span class='op'>)</span>


<span class='co'>##------------------------------------------------------------</span>
<span class='co'>## example of imputation in survival analysis</span>
<span class='co'>##------------------------------------------------------------</span>

<span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span><span class='op'>(</span><span class='va'>pbc</span>, package <span class='op'>=</span> <span class='st'>"randomForestSRC"</span><span class='op'>)</span>
<span class='va'>pbc.obj2</span> <span class='op'>&lt;-</span> <span class='fu'>rfsrc</span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/pkg/survival/man/Surv.html'>Surv</a></span><span class='op'>(</span><span class='va'>days</span>, <span class='va'>status</span><span class='op'>)</span> <span class='op'>~</span> <span class='va'>.</span>, <span class='va'>pbc</span>,
           nsplit <span class='op'>=</span> <span class='fl'>10</span>, na.action <span class='op'>=</span> <span class='st'>"na.impute"</span><span class='op'>)</span>


<span class='co'>## same as above but we iterate the missing data algorithm</span>
<span class='va'>pbc.obj3</span> <span class='op'>&lt;-</span> <span class='fu'>rfsrc</span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/pkg/survival/man/Surv.html'>Surv</a></span><span class='op'>(</span><span class='va'>days</span>, <span class='va'>status</span><span class='op'>)</span> <span class='op'>~</span> <span class='va'>.</span>, <span class='va'>pbc</span>,
         na.action <span class='op'>=</span> <span class='st'>"na.impute"</span>, nimpute <span class='op'>=</span> <span class='fl'>3</span><span class='op'>)</span>

<span class='co'>## fast way to impute the data (no inference is done)</span>
<span class='co'>## see impute for more details</span>
<span class='va'>pbc.imp</span> <span class='op'>&lt;-</span> <span class='fu'><a href='impute.rfsrc.html'>impute</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/pkg/survival/man/Surv.html'>Surv</a></span><span class='op'>(</span><span class='va'>days</span>, <span class='va'>status</span><span class='op'>)</span> <span class='op'>~</span> <span class='va'>.</span>, <span class='va'>pbc</span>, splitrule <span class='op'>=</span> <span class='st'>"random"</span><span class='op'>)</span>

<span class='co'>##------------------------------------------------------------</span>
<span class='co'>## compare RF-SRC to Cox regression</span>
<span class='co'>## Illustrates C-index and Brier score measures of performance</span>
<span class='co'>## assumes "pec" and "survival" libraries are loaded</span>
<span class='co'>##------------------------------------------------------------</span>

<span class='kw'>if</span> <span class='op'>(</span><span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='st'><a href='https://github.com/therneau/survival'>"survival"</a></span>, logical.return <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>
    <span class='op'>&amp;</span> <span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='st'>"pec"</span>, logical.return <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>
    <span class='op'>&amp;</span> <span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='st'>"prodlim"</span>, logical.return <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span><span class='op'>)</span>

<span class='op'>{</span>
  <span class='co'>##prediction function required for pec</span>
  <span class='va'>predictSurvProb.rfsrc</span> <span class='op'>&lt;-</span> <span class='kw'>function</span><span class='op'>(</span><span class='va'>object</span>, <span class='va'>newdata</span>, <span class='va'>times</span>, <span class='va'>...</span><span class='op'>)</span><span class='op'>{</span>
    <span class='va'>ptemp</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/stats/predict.html'>predict</a></span><span class='op'>(</span><span class='va'>object</span>,newdata<span class='op'>=</span><span class='va'>newdata</span>,<span class='va'>...</span><span class='op'>)</span><span class='op'>$</span><span class='va'>survival</span>
    <span class='va'>pos</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/prodlim/man/sindex.html'>sindex</a></span><span class='op'>(</span>jump.times <span class='op'>=</span> <span class='va'>object</span><span class='op'>$</span><span class='va'>time.interest</span>, eval.times <span class='op'>=</span> <span class='va'>times</span><span class='op'>)</span>
    <span class='va'>p</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/cbind.html'>cbind</a></span><span class='op'>(</span><span class='fl'>1</span>,<span class='va'>ptemp</span><span class='op'>)</span><span class='op'>[</span>, <span class='va'>pos</span> <span class='op'>+</span> <span class='fl'>1</span><span class='op'>]</span>
    <span class='kw'>if</span> <span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/nrow.html'>NROW</a></span><span class='op'>(</span><span class='va'>p</span><span class='op'>)</span> <span class='op'>!=</span> <span class='fu'><a href='https://rdrr.io/r/base/nrow.html'>NROW</a></span><span class='op'>(</span><span class='va'>newdata</span><span class='op'>)</span> <span class='op'>||</span> <span class='fu'><a href='https://rdrr.io/r/base/nrow.html'>NCOL</a></span><span class='op'>(</span><span class='va'>p</span><span class='op'>)</span> <span class='op'>!=</span> <span class='fu'><a href='https://rdrr.io/r/base/length.html'>length</a></span><span class='op'>(</span><span class='va'>times</span><span class='op'>)</span><span class='op'>)</span>
      <span class='kw'><a href='https://rdrr.io/r/base/stop.html'>stop</a></span><span class='op'>(</span><span class='st'>"Prediction failed"</span><span class='op'>)</span>
    <span class='va'>p</span>
  <span class='op'>}</span>

  <span class='co'>## data, formula specifications</span>
  <span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span><span class='op'>(</span><span class='va'>pbc</span>, package <span class='op'>=</span> <span class='st'>"randomForestSRC"</span><span class='op'>)</span>
  <span class='va'>pbc.na</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/stats/na.fail.html'>na.omit</a></span><span class='op'>(</span><span class='va'>pbc</span><span class='op'>)</span>  <span class='co'>##remove NA's</span>
  <span class='va'>surv.f</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/stats/formula.html'>as.formula</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/pkg/survival/man/Surv.html'>Surv</a></span><span class='op'>(</span><span class='va'>days</span>, <span class='va'>status</span><span class='op'>)</span> <span class='op'>~</span> <span class='va'>.</span><span class='op'>)</span>
  <span class='va'>pec.f</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/stats/formula.html'>as.formula</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/pkg/prodlim/man/Hist.html'>Hist</a></span><span class='op'>(</span><span class='va'>days</span>,<span class='va'>status</span><span class='op'>)</span> <span class='op'>~</span> <span class='fl'>1</span><span class='op'>)</span>

  <span class='co'>## run cox/rfsrc models</span>
  <span class='co'>## for illustration we use a small number of trees</span>
  <span class='va'>cox.obj</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/survival/man/coxph.html'>coxph</a></span><span class='op'>(</span><span class='va'>surv.f</span>, data <span class='op'>=</span> <span class='va'>pbc.na</span>, x <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>
  <span class='va'>rfsrc.obj</span> <span class='op'>&lt;-</span> <span class='fu'>rfsrc</span><span class='op'>(</span><span class='va'>surv.f</span>, <span class='va'>pbc.na</span>, ntree <span class='op'>=</span> <span class='fl'>150</span><span class='op'>)</span>

  <span class='co'>## compute bootstrap cross-validation estimate of expected Brier score</span>
  <span class='co'>## see Mogensen, Ishwaran and Gerds (2012) Journal of Statistical Software</span>
  <span class='fu'><a href='https://rdrr.io/r/base/Random.html'>set.seed</a></span><span class='op'>(</span><span class='fl'>17743</span><span class='op'>)</span>
  <span class='va'>prederror.pbc</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/pec/man/pec.html'>pec</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span><span class='va'>cox.obj</span>,<span class='va'>rfsrc.obj</span><span class='op'>)</span>, data <span class='op'>=</span> <span class='va'>pbc.na</span>, formula <span class='op'>=</span> <span class='va'>pec.f</span>,
                        splitMethod <span class='op'>=</span> <span class='st'>"bootcv"</span>, B <span class='op'>=</span> <span class='fl'>50</span><span class='op'>)</span>
  <span class='fu'><a href='https://rdrr.io/r/base/print.html'>print</a></span><span class='op'>(</span><span class='va'>prederror.pbc</span><span class='op'>)</span>
  <span class='fu'><a href='https://rdrr.io/r/graphics/plot.default.html'>plot</a></span><span class='op'>(</span><span class='va'>prederror.pbc</span><span class='op'>)</span>

  <span class='co'>## compute out-of-bag C-index for cox regression and compare to rfsrc</span>
  <span class='va'>rfsrc.obj</span> <span class='op'>&lt;-</span> <span class='fu'>rfsrc</span><span class='op'>(</span><span class='va'>surv.f</span>, <span class='va'>pbc.na</span><span class='op'>)</span>
  <span class='fu'><a href='https://rdrr.io/r/base/cat.html'>cat</a></span><span class='op'>(</span><span class='st'>"out-of-bag Cox Analysis ..."</span>, <span class='st'>"\n"</span><span class='op'>)</span>
  <span class='va'>cox.err</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/lapply.html'>sapply</a></span><span class='op'>(</span><span class='fl'>1</span><span class='op'>:</span><span class='fl'>100</span>, <span class='kw'>function</span><span class='op'>(</span><span class='va'>b</span><span class='op'>)</span> <span class='op'>{</span>
    <span class='kw'>if</span> <span class='op'>(</span><span class='va'>b</span><span class='op'>%%</span><span class='fl'>10</span> <span class='op'>==</span> <span class='fl'>0</span><span class='op'>)</span> <span class='fu'><a href='https://rdrr.io/r/base/cat.html'>cat</a></span><span class='op'>(</span><span class='st'>"cox bootstrap:"</span>, <span class='va'>b</span>, <span class='st'>"\n"</span><span class='op'>)</span>
    <span class='va'>train</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/sample.html'>sample</a></span><span class='op'>(</span><span class='fl'>1</span><span class='op'>:</span><span class='fu'><a href='https://rdrr.io/r/base/nrow.html'>nrow</a></span><span class='op'>(</span><span class='va'>pbc.na</span><span class='op'>)</span>, <span class='fu'><a href='https://rdrr.io/r/base/nrow.html'>nrow</a></span><span class='op'>(</span><span class='va'>pbc.na</span><span class='op'>)</span>, replace <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>
    <span class='va'>cox.obj</span> <span class='op'>&lt;-</span> <span class='kw'><a href='https://rdrr.io/r/base/conditions.html'>tryCatch</a></span><span class='op'>(</span><span class='op'>{</span><span class='fu'><a href='https://rdrr.io/pkg/survival/man/coxph.html'>coxph</a></span><span class='op'>(</span><span class='va'>surv.f</span>, <span class='va'>pbc.na</span><span class='op'>[</span><span class='va'>train</span>, <span class='op'>]</span><span class='op'>)</span><span class='op'>}</span>, error<span class='op'>=</span><span class='kw'>function</span><span class='op'>(</span><span class='va'>ex</span><span class='op'>)</span><span class='op'>{</span><span class='cn'>NULL</span><span class='op'>}</span><span class='op'>)</span>
    <span class='kw'>if</span> <span class='op'>(</span><span class='op'>!</span><span class='fu'><a href='https://rdrr.io/r/base/NULL.html'>is.null</a></span><span class='op'>(</span><span class='va'>cox.obj</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>{</span>
      <span class='fu'>get.cindex</span><span class='op'>(</span><span class='va'>pbc.na</span><span class='op'>$</span><span class='va'>days</span><span class='op'>[</span><span class='op'>-</span><span class='va'>train</span><span class='op'>]</span>, <span class='va'>pbc.na</span><span class='op'>$</span><span class='va'>status</span><span class='op'>[</span><span class='op'>-</span><span class='va'>train</span><span class='op'>]</span>, <span class='fu'><a href='https://rdrr.io/r/stats/predict.html'>predict</a></span><span class='op'>(</span><span class='va'>cox.obj</span>, <span class='va'>pbc.na</span><span class='op'>[</span><span class='op'>-</span><span class='va'>train</span>, <span class='op'>]</span><span class='op'>)</span><span class='op'>)</span>
    <span class='op'>}</span> <span class='kw'>else</span> <span class='cn'>NA</span>
  <span class='op'>}</span><span class='op'>)</span>
  <span class='fu'><a href='https://rdrr.io/r/base/cat.html'>cat</a></span><span class='op'>(</span><span class='st'>"\n\tOOB error rates\n\n"</span><span class='op'>)</span>
  <span class='fu'><a href='https://rdrr.io/r/base/cat.html'>cat</a></span><span class='op'>(</span><span class='st'>"\tRSF            : "</span>, <span class='va'>rfsrc.obj</span><span class='op'>$</span><span class='va'>err.rate</span><span class='op'>[</span><span class='va'>rfsrc.obj</span><span class='op'>$</span><span class='va'>ntree</span><span class='op'>]</span>, <span class='st'>"\n"</span><span class='op'>)</span>
  <span class='fu'><a href='https://rdrr.io/r/base/cat.html'>cat</a></span><span class='op'>(</span><span class='st'>"\tCox regression : "</span>, <span class='fu'><a href='https://rdrr.io/r/base/mean.html'>mean</a></span><span class='op'>(</span><span class='va'>cox.err</span>, na.rm <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>, <span class='st'>"\n"</span><span class='op'>)</span>
<span class='op'>}</span>

<span class='co'>##------------------------------------------------------------</span>
<span class='co'>## competing risks</span>
<span class='co'>##------------------------------------------------------------</span>

<span class='co'>## WIHS analysis</span>
<span class='co'>## cumulative incidence function (CIF) for HAART and AIDS stratified by IDU</span>

<span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span><span class='op'>(</span><span class='va'>wihs</span>, package <span class='op'>=</span> <span class='st'>"randomForestSRC"</span><span class='op'>)</span>
<span class='va'>wihs.obj</span> <span class='op'>&lt;-</span> <span class='fu'>rfsrc</span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/pkg/survival/man/Surv.html'>Surv</a></span><span class='op'>(</span><span class='va'>time</span>, <span class='va'>status</span><span class='op'>)</span> <span class='op'>~</span> <span class='va'>.</span>, <span class='va'>wihs</span>, nsplit <span class='op'>=</span> <span class='fl'>3</span>, ntree <span class='op'>=</span> <span class='fl'>100</span><span class='op'>)</span>
<span class='fu'><a href='plot.competing.risk.rfsrc.html'>plot.competing.risk</a></span><span class='op'>(</span><span class='va'>wihs.obj</span><span class='op'>)</span>
<span class='va'>cif</span> <span class='op'>&lt;-</span> <span class='va'>wihs.obj</span><span class='op'>$</span><span class='va'>cif.oob</span>
<span class='va'>Time</span> <span class='op'>&lt;-</span> <span class='va'>wihs.obj</span><span class='op'>$</span><span class='va'>time.interest</span>
<span class='va'>idu</span> <span class='op'>&lt;-</span> <span class='va'>wihs</span><span class='op'>$</span><span class='va'>idu</span>
<span class='va'>cif.haart</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/cbind.html'>cbind</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/apply.html'>apply</a></span><span class='op'>(</span><span class='va'>cif</span><span class='op'>[</span>,,<span class='fl'>1</span><span class='op'>]</span><span class='op'>[</span><span class='va'>idu</span> <span class='op'>==</span> <span class='fl'>0</span>,<span class='op'>]</span>, <span class='fl'>2</span>, <span class='va'>mean</span><span class='op'>)</span>,
                   <span class='fu'><a href='https://rdrr.io/r/base/apply.html'>apply</a></span><span class='op'>(</span><span class='va'>cif</span><span class='op'>[</span>,,<span class='fl'>1</span><span class='op'>]</span><span class='op'>[</span><span class='va'>idu</span> <span class='op'>==</span> <span class='fl'>1</span>,<span class='op'>]</span>, <span class='fl'>2</span>, <span class='va'>mean</span><span class='op'>)</span><span class='op'>)</span>
<span class='va'>cif.aids</span>  <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/cbind.html'>cbind</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/apply.html'>apply</a></span><span class='op'>(</span><span class='va'>cif</span><span class='op'>[</span>,,<span class='fl'>2</span><span class='op'>]</span><span class='op'>[</span><span class='va'>idu</span> <span class='op'>==</span> <span class='fl'>0</span>,<span class='op'>]</span>, <span class='fl'>2</span>, <span class='va'>mean</span><span class='op'>)</span>,
                   <span class='fu'><a href='https://rdrr.io/r/base/apply.html'>apply</a></span><span class='op'>(</span><span class='va'>cif</span><span class='op'>[</span>,,<span class='fl'>2</span><span class='op'>]</span><span class='op'>[</span><span class='va'>idu</span> <span class='op'>==</span> <span class='fl'>1</span>,<span class='op'>]</span>, <span class='fl'>2</span>, <span class='va'>mean</span><span class='op'>)</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/graphics/matplot.html'>matplot</a></span><span class='op'>(</span><span class='va'>Time</span>, <span class='fu'><a href='https://rdrr.io/r/base/cbind.html'>cbind</a></span><span class='op'>(</span><span class='va'>cif.haart</span>, <span class='va'>cif.aids</span><span class='op'>)</span>, type <span class='op'>=</span> <span class='st'>"l"</span>,
        lty <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>1</span>,<span class='fl'>2</span>,<span class='fl'>1</span>,<span class='fl'>2</span><span class='op'>)</span>, col <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>4</span>, <span class='fl'>4</span>, <span class='fl'>2</span>, <span class='fl'>2</span><span class='op'>)</span>, lwd <span class='op'>=</span> <span class='fl'>3</span>,
        ylab <span class='op'>=</span> <span class='st'>"Cumulative Incidence"</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/graphics/legend.html'>legend</a></span><span class='op'>(</span><span class='st'>"topleft"</span>,
       legend <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"HAART (Non-IDU)"</span>, <span class='st'>"HAART (IDU)"</span>, <span class='st'>"AIDS (Non-IDU)"</span>, <span class='st'>"AIDS (IDU)"</span><span class='op'>)</span>,
       lty <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>1</span>,<span class='fl'>2</span>,<span class='fl'>1</span>,<span class='fl'>2</span><span class='op'>)</span>, col <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>4</span>, <span class='fl'>4</span>, <span class='fl'>2</span>, <span class='fl'>2</span><span class='op'>)</span>, lwd <span class='op'>=</span> <span class='fl'>3</span>, cex <span class='op'>=</span> <span class='fl'>1.5</span><span class='op'>)</span>


<span class='co'>## illustrates the various splitting rules</span>
<span class='co'>## illustrates event specific and non-event specific variable selection</span>
<span class='kw'>if</span> <span class='op'>(</span><span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='st'><a href='https://github.com/therneau/survival'>"survival"</a></span>, logical.return <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>{</span>

  <span class='co'>## use the pbc data from the survival package</span>
  <span class='co'>## events are transplant (1) and death (2)</span>
  <span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span><span class='op'>(</span><span class='va'>pbc</span>, package <span class='op'>=</span> <span class='st'>"survival"</span><span class='op'>)</span>
  <span class='va'>pbc</span><span class='op'>$</span><span class='va'>id</span> <span class='op'>&lt;-</span> <span class='cn'>NULL</span>

  <span class='co'>## modified Gray's weighted log-rank splitting</span>
  <span class='co'>## (equivalent to cause=c(1,1) and splitrule="logrankCR")</span>
  <span class='va'>pbc.cr</span> <span class='op'>&lt;-</span> <span class='fu'>rfsrc</span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/pkg/survival/man/Surv.html'>Surv</a></span><span class='op'>(</span><span class='va'>time</span>, <span class='va'>status</span><span class='op'>)</span> <span class='op'>~</span> <span class='va'>.</span>, <span class='va'>pbc</span><span class='op'>)</span>
 
  <span class='co'>## log-rank cause-1 specific splitting and targeted VIMP for cause 1</span>
  <span class='va'>pbc.log1</span> <span class='op'>&lt;-</span> <span class='fu'>rfsrc</span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/pkg/survival/man/Surv.html'>Surv</a></span><span class='op'>(</span><span class='va'>time</span>, <span class='va'>status</span><span class='op'>)</span> <span class='op'>~</span> <span class='va'>.</span>, <span class='va'>pbc</span>, 
              splitrule <span class='op'>=</span> <span class='st'>"logrank"</span>, cause <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>1</span>,<span class='fl'>0</span><span class='op'>)</span>, importance <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>

  <span class='co'>## log-rank cause-2 specific splitting and targeted VIMP for cause 2</span>
  <span class='va'>pbc.log2</span> <span class='op'>&lt;-</span> <span class='fu'>rfsrc</span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/pkg/survival/man/Surv.html'>Surv</a></span><span class='op'>(</span><span class='va'>time</span>, <span class='va'>status</span><span class='op'>)</span> <span class='op'>~</span> <span class='va'>.</span>, <span class='va'>pbc</span>, 
              splitrule <span class='op'>=</span> <span class='st'>"logrank"</span>, cause <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>0</span>,<span class='fl'>1</span><span class='op'>)</span>, importance <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>

  <span class='co'>## extract VIMP from the log-rank forests: event-specific</span>
  <span class='co'>## extract minimal depth from the Gray log-rank forest: non-event specific</span>
  <span class='va'>var.perf</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/data.frame.html'>data.frame</a></span><span class='op'>(</span>md <span class='op'>=</span> <span class='fu'>max.subtree</span><span class='op'>(</span><span class='va'>pbc.cr</span><span class='op'>)</span><span class='op'>$</span><span class='va'>order</span><span class='op'>[</span>, <span class='fl'>1</span><span class='op'>]</span>,
                         vimp1 <span class='op'>=</span> <span class='fl'>100</span> <span class='op'>*</span> <span class='va'>pbc.log1</span><span class='op'>$</span><span class='va'>importance</span><span class='op'>[</span> ,<span class='fl'>1</span><span class='op'>]</span>,
                         vimp2 <span class='op'>=</span> <span class='fl'>100</span> <span class='op'>*</span> <span class='va'>pbc.log2</span><span class='op'>$</span><span class='va'>importance</span><span class='op'>[</span> ,<span class='fl'>2</span><span class='op'>]</span><span class='op'>)</span>
  <span class='fu'><a href='https://rdrr.io/r/base/print.html'>print</a></span><span class='op'>(</span><span class='va'>var.perf</span><span class='op'>[</span><span class='fu'><a href='https://rdrr.io/r/base/order.html'>order</a></span><span class='op'>(</span><span class='va'>var.perf</span><span class='op'>$</span><span class='va'>md</span><span class='op'>)</span>, <span class='op'>]</span>, digits <span class='op'>=</span> <span class='fl'>2</span><span class='op'>)</span>

<span class='op'>}</span>

<span class='co'>## ------------------------------------------------------------</span>
<span class='co'>## regression analysis</span>
<span class='co'>## ------------------------------------------------------------</span>

<span class='co'>## new York air quality measurements</span>
<span class='va'>airq.obj</span> <span class='op'>&lt;-</span> <span class='fu'>rfsrc</span><span class='op'>(</span><span class='va'>Ozone</span> <span class='op'>~</span> <span class='va'>.</span>, data <span class='op'>=</span> <span class='va'>airquality</span>, na.action <span class='op'>=</span> <span class='st'>"na.impute"</span><span class='op'>)</span>

<span class='co'># partial plot of variables (see plot.variable for more details)</span>
<span class='fu'><a href='plot.variable.rfsrc.html'>plot.variable</a></span><span class='op'>(</span><span class='va'>airq.obj</span>, partial <span class='op'>=</span> <span class='cn'>TRUE</span>, smooth.lines <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>

<span class='co'>## motor trend cars</span>
<span class='va'>mtcars.obj</span> <span class='op'>&lt;-</span> <span class='fu'>rfsrc</span><span class='op'>(</span><span class='va'>mpg</span> <span class='op'>~</span> <span class='va'>.</span>, data <span class='op'>=</span> <span class='va'>mtcars</span><span class='op'>)</span>

<span class='co'>## ------------------------------------------------------------</span>
<span class='co'>## regression with custom bootstrap</span>
<span class='co'>## ------------------------------------------------------------</span>

<span class='va'>ntree</span> <span class='op'>&lt;-</span> <span class='fl'>25</span>
<span class='va'>n</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/nrow.html'>nrow</a></span><span class='op'>(</span><span class='va'>mtcars</span><span class='op'>)</span>
<span class='va'>s.size</span> <span class='op'>&lt;-</span> <span class='va'>n</span> <span class='op'>/</span> <span class='fl'>2</span>
<span class='va'>swr</span> <span class='op'>&lt;-</span> <span class='cn'>TRUE</span>
<span class='va'>samp</span> <span class='op'>&lt;-</span> <span class='fu'>randomForestSRC</span><span class='fu'>:::</span><span class='fu'>make.sample</span><span class='op'>(</span><span class='va'>ntree</span>, <span class='va'>n</span>, <span class='va'>s.size</span>, <span class='va'>swr</span><span class='op'>)</span>
<span class='va'>o</span> <span class='op'>&lt;-</span> <span class='fu'>rfsrc</span><span class='op'>(</span><span class='va'>mpg</span> <span class='op'>~</span> <span class='va'>.</span>, <span class='va'>mtcars</span>, bootstrap <span class='op'>=</span> <span class='st'>"by.user"</span>, samp <span class='op'>=</span> <span class='va'>samp</span><span class='op'>)</span>

<span class='co'>## ------------------------------------------------------------</span>
<span class='co'>## classification analysis</span>
<span class='co'>## ------------------------------------------------------------</span>

<span class='co'>## iris data</span>
<span class='va'>iris.obj</span> <span class='op'>&lt;-</span> <span class='fu'>rfsrc</span><span class='op'>(</span><span class='va'>Species</span> <span class='op'>~</span><span class='va'>.</span>, data <span class='op'>=</span> <span class='va'>iris</span><span class='op'>)</span>

<span class='co'>## wisconsin prognostic breast cancer data</span>
<span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span><span class='op'>(</span><span class='va'>breast</span>, package <span class='op'>=</span> <span class='st'>"randomForestSRC"</span><span class='op'>)</span>
<span class='va'>breast.obj</span> <span class='op'>&lt;-</span> <span class='fu'>rfsrc</span><span class='op'>(</span><span class='va'>status</span> <span class='op'>~</span> <span class='va'>.</span>, data <span class='op'>=</span> <span class='va'>breast</span>, block.size<span class='op'>=</span><span class='fl'>1</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/graphics/plot.default.html'>plot</a></span><span class='op'>(</span><span class='va'>breast.obj</span><span class='op'>)</span>

<span class='co'>## ------------------------------------------------------------</span>
<span class='co'>## imbalanced classification data</span>
<span class='co'>## see the "imbalanced" function for further details</span>
<span class='co'>##</span>
<span class='co'>## a) use balanced random forests with undersampling of the majority class</span>
<span class='co'>## Specifically let n0, n1 be sample sizes for majority, minority</span>
<span class='co'>## cases.  We sample 2 x n1 cases with majority, minority cases chosen</span>
<span class='co'>## with probabilities n1/n, n0/n where n=n0+n1</span>
<span class='co'>##</span>
<span class='co'>## b) balanced random forests using "imbalanced"</span>
<span class='co'>##</span>
<span class='co'>## c) q-classifier (RFQ) using "imbalanced" </span>
<span class='co'>##</span>
<span class='co'>## ------------------------------------------------------------</span>

<span class='co'>## Wisconsin breast cancer example</span>
<span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span><span class='op'>(</span><span class='va'>breast</span>, package <span class='op'>=</span> <span class='st'>"randomForestSRC"</span><span class='op'>)</span>
<span class='va'>breast</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/stats/na.fail.html'>na.omit</a></span><span class='op'>(</span><span class='va'>breast</span><span class='op'>)</span>

<span class='co'>## balanced random forests - brute force</span>
<span class='va'>y</span> <span class='op'>&lt;-</span> <span class='va'>breast</span><span class='op'>$</span><span class='va'>status</span>
<span class='va'>obdirect</span> <span class='op'>&lt;-</span> <span class='fu'>rfsrc</span><span class='op'>(</span><span class='va'>status</span> <span class='op'>~</span> <span class='va'>.</span>, data <span class='op'>=</span> <span class='va'>breast</span>, nsplit <span class='op'>=</span> <span class='fl'>10</span>,
            case.wt <span class='op'>=</span> <span class='fu'>randomForestSRC</span><span class='fu'>:::</span><span class='fu'>make.wt</span><span class='op'>(</span><span class='va'>y</span><span class='op'>)</span>,
            sampsize <span class='op'>=</span> <span class='fu'>randomForestSRC</span><span class='fu'>:::</span><span class='fu'>make.size</span><span class='op'>(</span><span class='va'>y</span><span class='op'>)</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/print.html'>print</a></span><span class='op'>(</span><span class='va'>obdirect</span><span class='op'>)</span>

<span class='co'>## balanced random forests - using "imbalanced" </span>
<span class='va'>ob</span> <span class='op'>&lt;-</span> <span class='fu'><a href='imbalanced.rfsrc.html'>imbalanced</a></span><span class='op'>(</span><span class='va'>status</span> <span class='op'>~</span> <span class='va'>.</span>, data <span class='op'>=</span> <span class='va'>breast</span>, method <span class='op'>=</span> <span class='st'>"brf"</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/print.html'>print</a></span><span class='op'>(</span><span class='va'>ob</span><span class='op'>)</span>

<span class='co'>## q-classifier (RFQ) - using "imbalanced" </span>
<span class='va'>oq</span> <span class='op'>&lt;-</span> <span class='fu'><a href='imbalanced.rfsrc.html'>imbalanced</a></span><span class='op'>(</span><span class='va'>status</span> <span class='op'>~</span> <span class='va'>.</span>, data <span class='op'>=</span> <span class='va'>breast</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/print.html'>print</a></span><span class='op'>(</span><span class='va'>oq</span><span class='op'>)</span>

<span class='co'>## q-classifier (RFQ) - with auc splitting</span>
<span class='va'>oqauc</span> <span class='op'>&lt;-</span> <span class='fu'><a href='imbalanced.rfsrc.html'>imbalanced</a></span><span class='op'>(</span><span class='va'>status</span> <span class='op'>~</span> <span class='va'>.</span>, data <span class='op'>=</span> <span class='va'>breast</span>, splitrule <span class='op'>=</span> <span class='st'>"auc"</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/print.html'>print</a></span><span class='op'>(</span><span class='va'>oqauc</span><span class='op'>)</span>

<span class='co'>## ------------------------------------------------------------</span>
<span class='co'>## unsupervised analysis</span>
<span class='co'>## ------------------------------------------------------------</span>

<span class='co'>## two equivalent ways to implement unsupervised forests</span>
<span class='va'>mtcars.unspv</span> <span class='op'>&lt;-</span> <span class='fu'>rfsrc</span><span class='op'>(</span><span class='fu'>Unsupervised</span><span class='op'>(</span><span class='op'>)</span> <span class='op'>~</span><span class='va'>.</span>, data <span class='op'>=</span> <span class='va'>mtcars</span><span class='op'>)</span>
<span class='va'>mtcars2.unspv</span> <span class='op'>&lt;-</span> <span class='fu'>rfsrc</span><span class='op'>(</span>data <span class='op'>=</span> <span class='va'>mtcars</span><span class='op'>)</span>


<span class='co'>## illustration of sidClustering for the mtcars data</span>
<span class='co'>## see sidClustering for more details</span>
<span class='va'>mtcars.sid</span> <span class='op'>&lt;-</span> <span class='fu'><a href='sidClustering.rfsrc.html'>sidClustering</a></span><span class='op'>(</span><span class='va'>mtcars</span>, k <span class='op'>=</span> <span class='fl'>1</span><span class='op'>:</span><span class='fl'>10</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/print.html'>print</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/split.html'>split</a></span><span class='op'>(</span><span class='va'>mtcars</span>, <span class='va'>mtcars.sid</span><span class='op'>$</span><span class='va'>cl</span><span class='op'>[</span>, <span class='fl'>3</span><span class='op'>]</span><span class='op'>)</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/print.html'>print</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/split.html'>split</a></span><span class='op'>(</span><span class='va'>mtcars</span>, <span class='va'>mtcars.sid</span><span class='op'>$</span><span class='va'>cl</span><span class='op'>[</span>, <span class='fl'>10</span><span class='op'>]</span><span class='op'>)</span><span class='op'>)</span>


<span class='co'>## ------------------------------------------------------------</span>
<span class='co'>## multivariate regression analysis</span>
<span class='co'>## ------------------------------------------------------------</span>

<span class='va'>mtcars.mreg</span> <span class='op'>&lt;-</span> <span class='fu'>rfsrc</span><span class='op'>(</span><span class='fu'>Multivar</span><span class='op'>(</span><span class='va'>mpg</span>, <span class='va'>cyl</span><span class='op'>)</span> <span class='op'>~</span><span class='va'>.</span>, data <span class='op'>=</span> <span class='va'>mtcars</span>,
           block.size<span class='op'>=</span><span class='fl'>1</span>, importance <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>

<span class='co'>## extract error rates, vimp, and OOB predicted values for all targets</span>
<span class='va'>err</span> <span class='op'>&lt;-</span> <span class='fu'>get.mv.error</span><span class='op'>(</span><span class='va'>mtcars.mreg</span><span class='op'>)</span>
<span class='va'>vmp</span> <span class='op'>&lt;-</span> <span class='fu'>get.mv.vimp</span><span class='op'>(</span><span class='va'>mtcars.mreg</span><span class='op'>)</span>
<span class='va'>pred</span> <span class='op'>&lt;-</span> <span class='fu'>get.mv.predicted</span><span class='op'>(</span><span class='va'>mtcars.mreg</span><span class='op'>)</span>

<span class='co'>## standardized error and vimp</span>
<span class='va'>err.std</span> <span class='op'>&lt;-</span> <span class='fu'>get.mv.error</span><span class='op'>(</span><span class='va'>mtcars.mreg</span>, standardize <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>
<span class='va'>vmp.std</span> <span class='op'>&lt;-</span> <span class='fu'>get.mv.vimp</span><span class='op'>(</span><span class='va'>mtcars.mreg</span>, standardize <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>


<span class='co'>## ------------------------------------------------------------</span>
<span class='co'>## multivariate mixed forests (nutrigenomic study)</span>
<span class='co'>## study effects of diet, lipids and gene expression for mice</span>
<span class='co'>## diet, genotype and lipids used as the multivariate y</span>
<span class='co'>## genes used for the x features</span>
<span class='co'>## ------------------------------------------------------------</span>

<span class='co'>## load the data (data is a list)</span>
<span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span><span class='op'>(</span><span class='va'>nutrigenomic</span>, package <span class='op'>=</span> <span class='st'>"randomForestSRC"</span><span class='op'>)</span>

<span class='co'>## assemble the multivariate y data</span>
<span class='va'>ydta</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/data.frame.html'>data.frame</a></span><span class='op'>(</span>diet <span class='op'>=</span> <span class='va'>nutrigenomic</span><span class='op'>$</span><span class='va'>diet</span>,
                   genotype <span class='op'>=</span> <span class='va'>nutrigenomic</span><span class='op'>$</span><span class='va'>genotype</span>,
                   <span class='va'>nutrigenomic</span><span class='op'>$</span><span class='va'>lipids</span><span class='op'>)</span>

<span class='co'>## multivariate mixed forest call</span>
<span class='co'>## uses "get.mv.formula" for conveniently setting formula</span>
<span class='va'>mv.obj</span> <span class='op'>&lt;-</span> <span class='fu'>rfsrc</span><span class='op'>(</span><span class='fu'>get.mv.formula</span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>colnames</a></span><span class='op'>(</span><span class='va'>ydta</span><span class='op'>)</span><span class='op'>)</span>,
             <span class='fu'><a href='https://rdrr.io/r/base/data.frame.html'>data.frame</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/do.call.html'>do.call</a></span><span class='op'>(</span><span class='va'>cbind</span>, <span class='va'>nutrigenomic</span><span class='op'>)</span><span class='op'>)</span>,
             importance<span class='op'>=</span><span class='cn'>TRUE</span>, nsplit <span class='op'>=</span> <span class='fl'>10</span><span class='op'>)</span>

<span class='co'>## print results for diet and genotype y values      </span>
<span class='fu'><a href='https://rdrr.io/r/base/print.html'>print</a></span><span class='op'>(</span><span class='va'>mv.obj</span>, outcome.target <span class='op'>=</span> <span class='st'>"diet"</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/print.html'>print</a></span><span class='op'>(</span><span class='va'>mv.obj</span>, outcome.target <span class='op'>=</span> <span class='st'>"genotype"</span><span class='op'>)</span>

<span class='co'>## extract standardized VIMP</span>
<span class='va'>svimp</span> <span class='op'>&lt;-</span> <span class='fu'>get.mv.vimp</span><span class='op'>(</span><span class='va'>mv.obj</span>, standardize <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>

<span class='co'>## plot standardized VIMP for diet, genotype and lipid for each gene</span>
<span class='fu'><a href='https://rdrr.io/r/graphics/boxplot.html'>boxplot</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/t.html'>t</a></span><span class='op'>(</span><span class='va'>svimp</span><span class='op'>)</span>, col <span class='op'>=</span> <span class='st'>"bisque"</span>, cex.axis <span class='op'>=</span> <span class='fl'>.7</span>, las <span class='op'>=</span> <span class='fl'>2</span>,
        outline <span class='op'>=</span> <span class='cn'>FALSE</span>,
        ylab <span class='op'>=</span> <span class='st'>"standardized VIMP"</span>,
        main <span class='op'>=</span> <span class='st'>"diet/genotype/lipid VIMP for each gene"</span><span class='op'>)</span>
  
<span class='co'>## ------------------------------------------------------------</span>
<span class='co'>## custom splitting using the pre-coded examples</span>
<span class='co'>## ------------------------------------------------------------</span>

<span class='co'>## motor trend cars</span>
<span class='va'>mtcars.obj</span> <span class='op'>&lt;-</span> <span class='fu'>rfsrc</span><span class='op'>(</span><span class='va'>mpg</span> <span class='op'>~</span> <span class='va'>.</span>, data <span class='op'>=</span> <span class='va'>mtcars</span>, splitrule <span class='op'>=</span> <span class='st'>"custom"</span><span class='op'>)</span>

<span class='co'>## iris analysis</span>
<span class='va'>iris.obj</span> <span class='op'>&lt;-</span> <span class='fu'>rfsrc</span><span class='op'>(</span><span class='va'>Species</span> <span class='op'>~</span><span class='va'>.</span>, data <span class='op'>=</span> <span class='va'>iris</span>, splitrule <span class='op'>=</span> <span class='st'>"custom1"</span><span class='op'>)</span>

<span class='co'>## WIHS analysis</span>
<span class='va'>wihs.obj</span> <span class='op'>&lt;-</span> <span class='fu'>rfsrc</span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/pkg/survival/man/Surv.html'>Surv</a></span><span class='op'>(</span><span class='va'>time</span>, <span class='va'>status</span><span class='op'>)</span> <span class='op'>~</span> <span class='va'>.</span>, <span class='va'>wihs</span>, nsplit <span class='op'>=</span> <span class='fl'>3</span>,
                  ntree <span class='op'>=</span> <span class='fl'>100</span>, splitrule <span class='op'>=</span> <span class='st'>"custom1"</span><span class='op'>)</span>

<span class='co'># }</span>
</pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Min Lu.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


